{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0fd5ecf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "## Daniel Marten\n",
    "## GRCh37 Control Preparation\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "adb35230",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge and output for Liftover\n",
    "# list of intergenic non-orf and intergenic-orf dataframes\n",
    "norf_dfs = []\n",
    "orf_dfs = []\n",
    "\n",
    "for gtfidx in range(10):\n",
    "    # Reading in all GRCh37 ORFs from Nam\n",
    "    orf_path = f'/Users/marten/ug-gc/bam-gtf/grch37/sample-ORFs/GRCh37.Ens87.dna_rm.chromosome.all.intergenic_gt122.orfs.sample{gtfidx}.bed'\n",
    "    norf_path = f'/Users/marten/ug-gc/bam-gtf/grch37/sample-nORFs/GRCh37.Ens87.dna_rm.chromosome.all.intergenic_gt122.norf_sample{gtfidx}.bed'\n",
    "    new_orf = pd.read_csv(orf_path,sep='\\t').reset_index()\n",
    "    new_norf = pd.read_csv(norf_path,sep='\\t').reset_index()\n",
    "    \n",
    "    # Formatting - naming with _37\n",
    "    col_renames = {'level_0':'Chr',\n",
    "               'level_1':'Gene_Start_37',\n",
    "               'level_2':'Gene_End_37',\n",
    "               'level_3':'Name',\n",
    "               '# All coordinates are 0-indexed, coordinates in ID column are end-inclusive, and coordinates in start/end columns are end-exclusive (as is conventional)':'Strand'\n",
    "              }\n",
    "    \n",
    "    new_orf = new_orf.rename(columns=col_renames)\n",
    "    new_orf = new_orf.set_index('Name').drop('level_4',axis=1)\n",
    "    new_orf['Status'] = [f'orf_{gtfidx}']*new_orf.shape[0]\n",
    "    new_orf['Control_Set'] = (gtfidx//2)+1\n",
    "    #new_orf['annotation'] = ['intergenics']*new_orf.shape[0]\n",
    "    \n",
    "    new_norf = new_norf.rename(columns=col_renames)\n",
    "    new_norf = new_norf.set_index('Name').drop('level_4',axis=1)\n",
    "    new_norf['Status'] = [f'norf_{gtfidx}']*new_norf.shape[0]\n",
    "    new_norf['Control_Set'] = (gtfidx//2)+1\n",
    "    #new_norf['annotation'] = ['intergenics']*new_norf.shape[0]\n",
    "\n",
    "    orf_dfs.append(new_orf)\n",
    "    norf_dfs.append(new_norf)\n",
    "\n",
    "# Create a mega-dataframe with all orfs and non-orfs\n",
    "mega_37 = pd.concat(orf_dfs + norf_dfs)\n",
    "\n",
    "# code to turn the DFs as we have them formatted into properly formatted BEDs with the control set in their name\n",
    "# code copied from GRCh38 work, has old column names that we correct \n",
    "def bedify(input_df):\n",
    "    ret_df = input_df.reset_index()\n",
    "    newnames = [ret_df.loc[xi,'Name']+ret_df.loc[xi,'Status']+'_control_set_'+str(ret_df.loc[xi,'Control_Set']) for xi in ret_df.index] \n",
    "    ret_df['new_index'] = newnames\n",
    "    ret_df_2 = ret_df[['Chr','Gene_Start_37','Gene_End_37','new_index','Strand']]\n",
    "    ret_df_2['zeroes'] = 0\n",
    "    ret_df_2 = ret_df_2[['Chr','Gene_Start_37','Gene_End_37','new_index','zeroes','Strand']].sort_values(by=['Chr','Gene_Start_37','Gene_End_37'])\n",
    "    ret_df_2.new_index[1]\n",
    "    return ret_df_2\n",
    "\n",
    "threesevenbam = bedify(mega_37)\n",
    "\n",
    "# NOTICE: UNCOMMENT IF YOU NEED TO CREATE A MERGED BAM TO RUN THRU LIFTOVER \n",
    "# threesevenbam.to_csv(r'/Users/marten/ug-gc/merged_unrefined_37bam.bam',sep='\\t',index=False,header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c1ce541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_37</th>\n",
       "      <th>Gene_End_37</th>\n",
       "      <th>new_index</th>\n",
       "      <th>zeroes</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23269</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42671</td>\n",
       "      <td>42896</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23288</th>\n",
       "      <td>chr1</td>\n",
       "      <td>58253</td>\n",
       "      <td>58619</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20292</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59878</td>\n",
       "      <td>60067</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38957</th>\n",
       "      <td>chr1</td>\n",
       "      <td>105843</td>\n",
       "      <td>105990</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27000</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107280</td>\n",
       "      <td>107418</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59027971</td>\n",
       "      <td>59028151</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19989</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59209645</td>\n",
       "      <td>59209813</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59339711</td>\n",
       "      <td>59339870</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11991</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59353081</td>\n",
       "      <td>59353330</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15999</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59355200</td>\n",
       "      <td>59355356</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Gene_Start_37  Gene_End_37  \\\n",
       "23269  chr1          42671        42896   \n",
       "23288  chr1          58253        58619   \n",
       "20292  chr1          59878        60067   \n",
       "38957  chr1         105843       105990   \n",
       "27000  chr1         107280       107418   \n",
       "...     ...            ...          ...   \n",
       "1999   chrY       59027971     59028151   \n",
       "19989  chrY       59209645     59209813   \n",
       "1985   chrY       59339711     59339870   \n",
       "11991  chrY       59353081     59353330   \n",
       "15999  chrY       59355200     59355356   \n",
       "\n",
       "                                               new_index  zeroes Strand  \n",
       "23269  GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...       0      +  \n",
       "23288  GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...       0      -  \n",
       "20292  GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...       0      -  \n",
       "38957  GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...       0      +  \n",
       "27000  GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...       0      +  \n",
       "...                                                  ...     ...    ...  \n",
       "1999   GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "19989  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "1985   GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "11991  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "15999  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "\n",
       "[40000 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threesevenbam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ccca7c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take output above and run through UCSC's online liftover tool, with default settings \n",
    "# tool: https://genome.ucsc.edu/cgi-bin/hgLiftOver\n",
    "# and read back in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "07c408bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in liftover outputs and parse \n",
    "input_bam_liftedover = pd.read_csv(r'/Users/marten/Downloads/hglft_genome_2be94_2c4590.bed',sep='\\t',header=None,encoding=\"utf-8\").reset_index()\n",
    "# Formatting\n",
    "# Here, naming with _hg38 \n",
    "col_renames = {0:'Chr',\n",
    "           1:'Gene_Start_hg38',\n",
    "           2:'Gene_End_hg38',\n",
    "           3:'Name',\n",
    "           4:'Inclusivity',\n",
    "           5:'Strand'\n",
    "          }\n",
    "input_bam_liftedover = input_bam_liftedover.rename(columns=col_renames)\n",
    "input_bam_liftedover = input_bam_liftedover[['Chr','Gene_Start_hg38','Gene_End_hg38','Name','Inclusivity','Strand']]#.reset_index(by='Name')\n",
    "input_bam_liftedover.Gene_Start_hg38 = input_bam_liftedover.Gene_Start_hg38# - input_bam_liftedover.Inclusivity\n",
    "input_bam_liftedover.set_index('Name',inplace=True)\n",
    "input_bam_liftedover['Control_Set'] = [relevant_str.split('_')[-1] for relevant_str in input_bam_liftedover.index]\n",
    "input_bam_liftedover['Status'] = [relevant_str.split(':')[-1].split('_')[0][1:] for relevant_str in input_bam_liftedover.index]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "097c2f1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:132308-132529:-orf_7_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132308</td>\n",
       "      <td>132530</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr10.intergenic_gt122.orf:71432402-71432536:-orf_2_control_set_2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>69672646</td>\n",
       "      <td>69672781</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr12.intergenic_gt122.orf:125207652-125207828:+orf_1_control_set_1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>124723106</td>\n",
       "      <td>124723283</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr15.intergenic_gt122.orf:80607539-80607859:+orf_3_control_set_2</th>\n",
       "      <td>chr15</td>\n",
       "      <td>80315197</td>\n",
       "      <td>80315518</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr19.intergenic_gt122.orf:30248465-30248638:-orf_2_control_set_2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>29757558</td>\n",
       "      <td>29757732</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr2.intergenic_gt122.orf:240639420-240639593:+orf_2_control_set_2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>239717726</td>\n",
       "      <td>239717900</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr3.intergenic_gt122.orf:189073255-189073395:+orf_6_control_set_4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>189355466</td>\n",
       "      <td>189355607</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr5.intergenic_gt122.orf:89146277-89146447:-orf_0_control_set_1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>89850460</td>\n",
       "      <td>89850631</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr7.intergenic_gt122.orf:27256960-27257130:+orf_1_control_set_1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>27217341</td>\n",
       "      <td>27217512</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr9.intergenic_gt122.orf:8097091-8097258:+orf_2_control_set_2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>8097091</td>\n",
       "      <td>8097259</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...   chr1           132308   \n",
       "GRCh37.Ens87.dna_rm.chr10.intergenic_gt122.orf:...  chr10         69672646   \n",
       "GRCh37.Ens87.dna_rm.chr12.intergenic_gt122.orf:...  chr12        124723106   \n",
       "GRCh37.Ens87.dna_rm.chr15.intergenic_gt122.orf:...  chr15         80315197   \n",
       "GRCh37.Ens87.dna_rm.chr19.intergenic_gt122.orf:...  chr19         29757558   \n",
       "GRCh37.Ens87.dna_rm.chr2.intergenic_gt122.orf:2...   chr2        239717726   \n",
       "GRCh37.Ens87.dna_rm.chr3.intergenic_gt122.orf:1...   chr3        189355466   \n",
       "GRCh37.Ens87.dna_rm.chr5.intergenic_gt122.orf:8...   chr5         89850460   \n",
       "GRCh37.Ens87.dna_rm.chr7.intergenic_gt122.orf:2...   chr7         27217341   \n",
       "GRCh37.Ens87.dna_rm.chr9.intergenic_gt122.orf:8...   chr9          8097091   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...         132530   \n",
       "GRCh37.Ens87.dna_rm.chr10.intergenic_gt122.orf:...       69672781   \n",
       "GRCh37.Ens87.dna_rm.chr12.intergenic_gt122.orf:...      124723283   \n",
       "GRCh37.Ens87.dna_rm.chr15.intergenic_gt122.orf:...       80315518   \n",
       "GRCh37.Ens87.dna_rm.chr19.intergenic_gt122.orf:...       29757732   \n",
       "GRCh37.Ens87.dna_rm.chr2.intergenic_gt122.orf:2...      239717900   \n",
       "GRCh37.Ens87.dna_rm.chr3.intergenic_gt122.orf:1...      189355607   \n",
       "GRCh37.Ens87.dna_rm.chr5.intergenic_gt122.orf:8...       89850631   \n",
       "GRCh37.Ens87.dna_rm.chr7.intergenic_gt122.orf:2...       27217512   \n",
       "GRCh37.Ens87.dna_rm.chr9.intergenic_gt122.orf:8...        8097259   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr10.intergenic_gt122.orf:...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr12.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr15.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr19.intergenic_gt122.orf:...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr2.intergenic_gt122.orf:2...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr3.intergenic_gt122.orf:1...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr5.intergenic_gt122.orf:8...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr7.intergenic_gt122.orf:2...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr9.intergenic_gt122.orf:8...            1      +   \n",
       "\n",
       "                                                   Control_Set Status  \n",
       "Name                                                                   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...           4    orf  \n",
       "GRCh37.Ens87.dna_rm.chr10.intergenic_gt122.orf:...           2    orf  \n",
       "GRCh37.Ens87.dna_rm.chr12.intergenic_gt122.orf:...           1    orf  \n",
       "GRCh37.Ens87.dna_rm.chr15.intergenic_gt122.orf:...           2    orf  \n",
       "GRCh37.Ens87.dna_rm.chr19.intergenic_gt122.orf:...           2    orf  \n",
       "GRCh37.Ens87.dna_rm.chr2.intergenic_gt122.orf:2...           2    orf  \n",
       "GRCh37.Ens87.dna_rm.chr3.intergenic_gt122.orf:1...           4    orf  \n",
       "GRCh37.Ens87.dna_rm.chr5.intergenic_gt122.orf:8...           1    orf  \n",
       "GRCh37.Ens87.dna_rm.chr7.intergenic_gt122.orf:2...           1    orf  \n",
       "GRCh37.Ens87.dna_rm.chr9.intergenic_gt122.orf:8...           2    orf  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# input_bam_liftedover.Gene_Start_hg38 = input_bam_liftedover.Gene_Start_hg38-1\n",
    "input_bam_liftedover[input_bam_liftedover.Status=='orf'].iloc[::2001]\n",
    "# show (20000/2001=about 10) ORFs at random, from a mix of Chrs \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c34addf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:132308-132529:-orf_7_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>132308</td>\n",
       "      <td>132530</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:266925-267119:+orf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>297174</td>\n",
       "      <td>297369</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:330011-330244:-orf_6_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>489091</td>\n",
       "      <td>489325</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:340419-340610:-orf_8_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>478725</td>\n",
       "      <td>478917</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:378478-378666:+orf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>440669</td>\n",
       "      <td>440858</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59027971-59028150:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59209645-59209812:+orf_9_control_set_5</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57063496</td>\n",
       "      <td>57063664</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59339711-59339869:+orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57193560</td>\n",
       "      <td>57193719</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59353081-59353329:+orf_5_control_set_3</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57206930</td>\n",
       "      <td>57207179</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59355200-59355355:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19978 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...  chr1           132308   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:2...  chr1           297174   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...  chr1           489091   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...  chr1           478725   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...  chr1           440669   \n",
       "...                                                  ...              ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57063496   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57193560   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57206930   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...         132530   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:2...         297369   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...         489325   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...         478917   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...         440858   \n",
       "...                                                           ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       56882004   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57063664   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57193719   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57207179   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57209205   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:2...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...            1      -   \n",
       "...                                                         ...    ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      -   \n",
       "\n",
       "                                                   Control_Set Status  \n",
       "Name                                                                   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:1...           4    orf  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:2...           1    orf  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...           4    orf  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...           5    orf  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:3...           1    orf  \n",
       "...                                                        ...    ...  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           1    orf  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           5    orf  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           1    orf  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           3    orf  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           4    orf  \n",
       "\n",
       "[19978 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make a copy of liftedover controls \n",
    "control_df = input_bam_liftedover.copy()\n",
    "# Print ORFs in here \n",
    "control_df[control_df.Status=='orf']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ebeda0ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504-43241.norf_segment:42671-42895:+norf_1_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42671</td>\n",
       "      <td>42896</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>225</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:58253-58618:-norf_1_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>58253</td>\n",
       "      <td>58619</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>366</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208-60377.norf_segment:59878-60066:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59878</td>\n",
       "      <td>60067</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>189</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:105742-106428.norf_segment:105843-105989:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>105843</td>\n",
       "      <td>105990</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>norf</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:106531-107479.norf_segment:107280-107417:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107280</td>\n",
       "      <td>107418</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>138</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59027971-59028150:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>180</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59209645-59209812:+orf_9_control_set_5</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57063496</td>\n",
       "      <td>57063664</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "      <td>168</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59339711-59339869:+orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57193560</td>\n",
       "      <td>57193719</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>159</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59353081-59353329:+orf_5_control_set_3</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57206930</td>\n",
       "      <td>57207179</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>orf</td>\n",
       "      <td>249</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59355200-59355355:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>156</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39958 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...  chr1            42671   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...  chr1            58253   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...  chr1            59878   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...  chr1           105843   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...  chr1           107280   \n",
       "...                                                  ...              ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57063496   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57193560   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57206930   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...          42896   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...          58619   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...          60067   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...         105990   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...         107418   \n",
       "...                                                           ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       56882004   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57063664   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57193719   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57207179   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57209205   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...            1      +   \n",
       "...                                                         ...    ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      -   \n",
       "\n",
       "                                                   Control_Set Status  \\\n",
       "Name                                                                    \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...           5   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...           2   norf   \n",
       "...                                                        ...    ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           1    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           5    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           1    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           3    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           4    orf   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...          225    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...          366    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...          189    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...          147    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...          138    False   \n",
       "...                                                         ...      ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          180    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          168    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          159    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          249    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          156    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...          4  \n",
       "...                                                       ...  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "\n",
       "[39958 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add annotations for sorting later\n",
    "control_df['user_length'] = control_df.Gene_End_hg38 - control_df.Gene_Start_hg38\n",
    "control_df['removal'] = False\n",
    "control_df['heirarchy'] = 3\n",
    "control_df.loc[control_df['Status'].str.contains('norf'),'heirarchy'] = 4\n",
    "control_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "756d9eec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr14_GL000009v2_random\n",
      "chr15\n",
      "chr15_KI270850v1_alt\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr22_KI270879v1_alt\n",
      "chr3\n",
      "chr4\n",
      "chr4_GL000008v2_random\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr7_KI270803v1_alt\n",
      "chr8\n",
      "chr8_KI270821v1_alt\n",
      "chr9\n",
      "chrUn_KI270742v1\n",
      "chrX\n",
      "chrY\n",
      "Total duplicates removed (instances):  113 (note that first occurence is kept)\n",
      "Total non-duplicate overlaps observed:  497\n",
      "Overlaps decided by heirarchy:  250\n",
      "Overlaps decided by length:  242\n",
      "Overlaps where the first occuring one is taken:  5\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(39354, 10)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CODE TO: \n",
    "# by chromosome, in GRCh38 coordinates, remove controls that: \n",
    "# are duplicates of other controls\n",
    "# overlap other controls\n",
    "# decided by: \n",
    "# -> prioritize ORFs over Non-ORFs\n",
    "# -> prioritize controls which are longer\n",
    "# -> prioritize controls which start earlier\n",
    "# -> in the case of duplicates, keep one in earliest control set \n",
    "\n",
    "fixed_list = []\n",
    "dupl_list = []\n",
    "removed_list = []\n",
    "gi = 'Gene_Start_hg38'\n",
    "ge = 'Gene_End_hg38'\n",
    "\n",
    "total_dups = 0\n",
    "total_overlaps = 0\n",
    "by_heirarchy = 0\n",
    "by_length = 0\n",
    "punt = 0\n",
    "\n",
    "for unique_chrom in sorted(control_df.Chr.unique()):\n",
    "    print(unique_chrom)\n",
    "    by_chrom_df = control_df[control_df.Chr==unique_chrom]\n",
    "    \n",
    "    by_chrom_df = by_chrom_df.sort_values(by=['Chr',gi,ge])\n",
    "    \n",
    "    dupls = by_chrom_df[by_chrom_df.duplicated(['Chr','Gene_Start_hg38','Gene_End_hg38'],keep='first')]\n",
    "    dupl_list.append(dupls)\n",
    "    \n",
    "    by_chrom_df_dd = by_chrom_df[~by_chrom_df.duplicated(['Chr','Gene_Start_hg38','Gene_End_hg38'],keep='first')]\n",
    "    \n",
    "    total_dups += dupls.shape[0]\n",
    "    \n",
    "    for xi,yi in by_chrom_df_dd.iterrows():\n",
    "        \n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "        \n",
    "        query_new = by_chrom_df_dd[(by_chrom_df_dd[gi]>(index_start-1))&(by_chrom_df_dd[ge]<(index_end+1000000))]\n",
    "        \n",
    "        for xii,yii in query_new.iterrows():\n",
    "            \n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "            \n",
    "            case_1 = query_start in range(index_start,index_end+1)\n",
    "            case_2 = query_end in range(index_start,index_end+1)\n",
    "            case_3 = index_start in range(int(query_start),int(query_end)+1)\n",
    "            \n",
    "            if (xi!=xii):\n",
    "                if any([case_1,case_2,case_3]):\n",
    "                    # instance of overlap!\n",
    "                        total_overlaps += 1\n",
    "                        if yii['heirarchy'] > yi['heirarchy']:\n",
    "                            by_chrom_df_dd.loc[xii,'removal'] = True\n",
    "                            by_heirarchy += 1\n",
    "                        elif yii['heirarchy'] < yi['heirarchy']:\n",
    "                            by_chrom_df_dd.loc[xi,'removal'] = True\n",
    "                            by_heirarchy += 1\n",
    "                        else:\n",
    "                            if yii['user_length']>yi['user_length']:\n",
    "                                by_chrom_df_dd.loc[xi,'removal'] = True\n",
    "                                by_length += 1\n",
    "                            elif yii['user_length']<yi['user_length']:\n",
    "                                by_chrom_df_dd.loc[xii,'removal'] = True\n",
    "                                by_length += 1\n",
    "                            else:\n",
    "                                punt += 1\n",
    "                                if yii[gi] > yi[gi]:\n",
    "                                    by_chrom_df_dd.loc[xii,'removal'] = True\n",
    "                                else:\n",
    "                                    by_chrom_df_dd.loc[xi,'removal'] = True\n",
    "     \n",
    "    removed_list.append(by_chrom_df_dd[by_chrom_df_dd.removal])\n",
    "    by_chrom_df_dd = by_chrom_df_dd[~by_chrom_df_dd.removal]\n",
    "    fixed_list.append(by_chrom_df_dd)\n",
    "    \n",
    "print('Total duplicates removed (instances): ',total_dups, '(note that first occurence is kept)') # note that the first occurence is kept \n",
    "print('Total non-duplicate overlaps observed: ',total_overlaps)\n",
    "print('Overlaps decided by heirarchy: ',by_heirarchy)\n",
    "print('Overlaps decided by length: ',by_length)\n",
    "print('Overlaps where the first occuring one is taken: ',punt)\n",
    "\n",
    "fixed_df = pd.concat(fixed_list)\n",
    "\n",
    "fixed_df.shape\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe736b6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:65654142-65654270:+orf_7_control_set_4</th>\n",
       "      <td>chr14</td>\n",
       "      <td>65187424</td>\n",
       "      <td>65187553</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>129</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:72268066-72268188:+orf_7_control_set_4</th>\n",
       "      <td>chr14</td>\n",
       "      <td>71801349</td>\n",
       "      <td>71801472</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>123</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:99413699-99413845:+orf_6_control_set_4</th>\n",
       "      <td>chr14</td>\n",
       "      <td>98947362</td>\n",
       "      <td>98947509</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:104347821-104348036:+orf_7_control_set_4</th>\n",
       "      <td>chr14</td>\n",
       "      <td>103881484</td>\n",
       "      <td>103881700</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>216</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:104712926-104713135:+orf_8_control_set_5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>104246589</td>\n",
       "      <td>104246799</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "      <td>210</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:105970271-105970669:+orf_4_control_set_3</th>\n",
       "      <td>chr14</td>\n",
       "      <td>105503934</td>\n",
       "      <td>105504333</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>orf</td>\n",
       "      <td>399</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14         65187424   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14         71801349   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14         98947362   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14        103881484   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14        104246589   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14        105503934   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...       65187553   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...       71801472   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...       98947509   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...      103881700   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...      104246799   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...      105504333   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "\n",
       "                                                   Control_Set Status  \\\n",
       "Name                                                                    \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           4    orf   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           4    orf   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           4    orf   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           4    orf   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           5    orf   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           3    orf   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          129    False   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          123    False   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          147    False   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          216    False   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          210    False   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          399    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SANITY CHECK: looking at duplicates and making sure they're removed\n",
    "dupl_list[5] # \"random\" removed duplicates by chromosome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "01a20bb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:99413699-99413845:+orf_4_control_set_3</th>\n",
       "      <td>chr14</td>\n",
       "      <td>98947362</td>\n",
       "      <td>98947509</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>orf</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:99413699-99413845:+orf_6_control_set_4</th>\n",
       "      <td>chr14</td>\n",
       "      <td>98947362</td>\n",
       "      <td>98947509</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14         98947362   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14         98947362   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...       98947509   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...       98947509   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "\n",
       "                                                   Control_Set Status  \\\n",
       "Name                                                                    \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           3    orf   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           4    orf   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          147    False   \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          147    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how the duplicate looks in original\n",
    "control_df[(control_df.Chr=='chr14') & (control_df.Gene_Start_hg38==98947362)] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "06c2322d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:99413699-99413845:+orf_4_control_set_3</th>\n",
       "      <td>chr14</td>\n",
       "      <td>98947362</td>\n",
       "      <td>98947509</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>orf</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...  chr14         98947362   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...       98947509   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...            1      +   \n",
       "\n",
       "                                                   Control_Set Status  \\\n",
       "Name                                                                    \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...           3    orf   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          147    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh37.Ens87.dna_rm.chr14.intergenic_gt122.orf:...          3  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how the duplicate looks in final , see how there is only one entry \n",
    "fixed_df[(fixed_df.Chr=='chr14') & (fixed_df.Gene_Start_hg38==98947362)] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee090613",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:14920064-14921545.norf_segment:14921196-14921342:+norf_3_control_set_2</th>\n",
       "      <td>chr21</td>\n",
       "      <td>13548875</td>\n",
       "      <td>13549022</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>147</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:19923431-19924669.norf_segment:19924300-19924443:+norf_1_control_set_1</th>\n",
       "      <td>chr21</td>\n",
       "      <td>18551982</td>\n",
       "      <td>18552126</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>144</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:20137824-20139495.norf_segment:20139219-20139374:+norf_5_control_set_3</th>\n",
       "      <td>chr21</td>\n",
       "      <td>18766901</td>\n",
       "      <td>18767057</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>norf</td>\n",
       "      <td>156</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:21039616-21040413.norf_segment:21039952-21040134:-norf_3_control_set_2</th>\n",
       "      <td>chr21</td>\n",
       "      <td>19667638</td>\n",
       "      <td>19667821</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>183</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:28094960-28095651.norf_segment:28095175-28095432:-norf_7_control_set_4</th>\n",
       "      <td>chr21</td>\n",
       "      <td>26722856</td>\n",
       "      <td>26723114</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>norf</td>\n",
       "      <td>258</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:31462941-31463650.norf_segment:31463307-31463453:-norf_1_control_set_1</th>\n",
       "      <td>chr21</td>\n",
       "      <td>30090989</td>\n",
       "      <td>30091136</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>147</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:32173865-32175274.norf_segment:32174777-32174986:-norf_3_control_set_2</th>\n",
       "      <td>chr21</td>\n",
       "      <td>30802459</td>\n",
       "      <td>30802669</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>210</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:37390053-37391007.norf_segment:37390416-37390718:+norf_5_control_set_3</th>\n",
       "      <td>chr21</td>\n",
       "      <td>36018118</td>\n",
       "      <td>36018421</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>norf</td>\n",
       "      <td>303</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:43031348-43033576.norf_segment:43032333-43032497:+norf_5_control_set_3</th>\n",
       "      <td>chr21</td>\n",
       "      <td>41612173</td>\n",
       "      <td>41612338</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>norf</td>\n",
       "      <td>165</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1492...  chr21         13548875   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1992...  chr21         18551982   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2013...  chr21         18766901   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2103...  chr21         19667638   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...  chr21         26722856   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3146...  chr21         30090989   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3217...  chr21         30802459   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3739...  chr21         36018118   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:4303...  chr21         41612173   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1492...       13549022   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1992...       18552126   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2013...       18767057   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2103...       19667821   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...       26723114   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3146...       30091136   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3217...       30802669   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3739...       36018421   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:4303...       41612338   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1492...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1992...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2013...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2103...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3146...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3217...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3739...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:4303...            1      +   \n",
       "\n",
       "                                                   Control_Set Status  \\\n",
       "Name                                                                    \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1492...           2   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1992...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2013...           3   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2103...           2   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...           4   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3146...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3217...           2   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3739...           3   norf   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:4303...           3   norf   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1492...          147     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1992...          144     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2013...          156     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2103...          183     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...          258     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3146...          147     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3217...          210     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3739...          303     True   \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:4303...          165     True   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1492...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:1992...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2013...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2103...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3146...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3217...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:3739...          4  \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:4303...          4  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SANITY CHECK - making sure removed things are actually removed \n",
    "removed_list[len(removed_list)//2] # chosen chromosome of removed samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "826d98e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:28094960-28095651.norf_segment:28095175-28095432:-norf_7_control_set_4</th>\n",
       "      <td>chr21</td>\n",
       "      <td>26722856</td>\n",
       "      <td>26723114</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>norf</td>\n",
       "      <td>258</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...  chr21         26722856   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...       26723114   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...            1      -   \n",
       "\n",
       "                                                   Control_Set Status  \\\n",
       "Name                                                                    \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...           4   norf   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...          258    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh37.Ens87.dna_rm.chr21.intergenic_gt122:2809...          4  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# further sanity check at one of these locations to ensure only one entry \n",
    "control_df[(control_df.Chr=='chr21') & (control_df.Gene_Start_hg38==26722856)] # sequences removed \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c26ac8f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chr, Gene_Start_hg38, Gene_End_hg38, Inclusivity, Strand, Control_Set, Status, user_length, removal, heirarchy]\n",
       "Index: []"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking to make sure a random sequence is actually removed\n",
    "# REMOVED becuase it overlapped a higher priority sequence! not because it is a duplicate\n",
    "fixed_df[(fixed_df.Chr=='chr21') & (fixed_df.Gene_Start_hg38==26722856)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a1380488",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr14_GL000009v2_random\n",
      "chr15\n",
      "chr15_KI270850v1_alt\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr22_KI270879v1_alt\n",
      "chr3\n",
      "chr4\n",
      "chr4_GL000008v2_random\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr7_KI270803v1_alt\n",
      "chr8\n",
      "chr8_KI270821v1_alt\n",
      "chr9\n",
      "chrUn_KI270742v1\n",
      "chrX\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "# Check using the same code: \n",
    "gi = 'Gene_Start_hg38'\n",
    "ge = 'Gene_End_hg38'\n",
    "    \n",
    "for chrom_id in sorted(fixed_df.Chr.unique()):\n",
    "\n",
    "    print(chrom_id)\n",
    "\n",
    "    query_df = fixed_df[fixed_df['Chr']==chrom_id]\n",
    "\n",
    "    for xi,yi in query_df.iterrows():\n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "\n",
    "        query_new = query_df[(query_df[gi]>(index_start-1000000))&(query_df[ge]<(index_end+1000000))]\n",
    "\n",
    "        for xii,yii in query_new.iterrows():\n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "\n",
    "            if xi!=xii:\n",
    "                if query_start in range(index_start,index_end+1) or query_end in range(index_start,index_end+1):\n",
    "                    print(xi,yi)\n",
    "                    print(xii,yii) \n",
    "                    print('Exception(DANGER)')\n",
    "\n",
    "\n",
    "\n",
    "                elif index_start in range(int(query_start),int(query_end)+1):\n",
    "                    print(xi,yi)\n",
    "                    print(xii,yii) \n",
    "                    print('Exception(DANGER)')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c8b93040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PS</th>\n",
       "      <th>Description</th>\n",
       "      <th>Plength</th>\n",
       "      <th>Gap_Gene?</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>CDS_Start_hg38</th>\n",
       "      <th>CDS_End_hg38</th>\n",
       "      <th>Protein_Sequence</th>\n",
       "      <th>CDS_Sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_1</th>\n",
       "      <td>31</td>\n",
       "      <td>ENST00000308604.5_18272_457</td>\n",
       "      <td>52</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>111429413</td>\n",
       "      <td>111429572</td>\n",
       "      <td>MTDTENHDSSPSSTSTCCPPITAGMQLKDSLGPGSNCPLWTLRPLH...</td>\n",
       "      <td>ATGACAGACACTGAAAATCACGACTCATCCCCCTCCAGCACCTCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_10</th>\n",
       "      <td>31</td>\n",
       "      <td>ENST00000411630.2_23991_594</td>\n",
       "      <td>59</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>52713673</td>\n",
       "      <td>52713853</td>\n",
       "      <td>MLVATGQCSRCFMFTFSTFSFNCHNSEVDSVRDRLPQDHSAPANSM...</td>\n",
       "      <td>ATGCTGGTGGCAACAGGGCAGTGTAGCAGGTGCTTCATGTTCACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_103</th>\n",
       "      <td>31</td>\n",
       "      <td>ENST00000499346.2_27384_333</td>\n",
       "      <td>84</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>128082767</td>\n",
       "      <td>128083022</td>\n",
       "      <td>MLGAFRSGPQPLPEPRARCVPQPGLLWALTRRRESPLVTPGLNLEE...</td>\n",
       "      <td>ATGCTGGGGGCTTTCCGGTCGGGGCCGCAGCCGCTTCCGGAGCCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_108</th>\n",
       "      <td>24</td>\n",
       "      <td>ENST00000501177.3_12701_390</td>\n",
       "      <td>84</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>54919086</td>\n",
       "      <td>54925136</td>\n",
       "      <td>MLAEIHPKAGLQSLQFIMELLYWLLEGGDSEDKEDATGNVEMKNIQ...</td>\n",
       "      <td>ATGTTGGCTGAAATTCATCCCAAGGCTGGTCTGCAAAGTCTGCAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_112</th>\n",
       "      <td>25</td>\n",
       "      <td>ENST00000503704.1_24073_293</td>\n",
       "      <td>52</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>82900166</td>\n",
       "      <td>82900435</td>\n",
       "      <td>MRSREAGPKLRRIQEPANGSPGAVSETGGYREERLSDAEIMGKLLA...</td>\n",
       "      <td>ATGCGAAGCAGAGAGGCAGGACCAAAATTGAGGCGAATCCAGGAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_994</th>\n",
       "      <td>31</td>\n",
       "      <td>HF584391</td>\n",
       "      <td>69</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>143628143</td>\n",
       "      <td>143628353</td>\n",
       "      <td>MLYTHNTEFNLKRQICFVPQCKTFVSLCFVKQTQENWYTCTSWVLY...</td>\n",
       "      <td>ATGCTTTATACACATAATACTGAATTTAACCTCAAGAGGCAAATCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_995</th>\n",
       "      <td>31</td>\n",
       "      <td>HF583960</td>\n",
       "      <td>46</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>104020518</td>\n",
       "      <td>104021665</td>\n",
       "      <td>MREWLSIRNMRIKCEIFSCSVKPMSANCISCRMKNATCWLSMRLRN</td>\n",
       "      <td>ATGAGAGAATGGCTCAGCATCAGAAACATGAGAATCAAATGCGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_997</th>\n",
       "      <td>31</td>\n",
       "      <td>HF548108</td>\n",
       "      <td>40</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>70860637</td>\n",
       "      <td>70860760</td>\n",
       "      <td>MFAYKGSSYHVSNTSNSINPTPKLASNPVGRYCMIKCLII</td>\n",
       "      <td>ATGTTTGCATATAAGGGAAGTAGTTATCATGTTAGTAATACCTCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_998</th>\n",
       "      <td>31</td>\n",
       "      <td>HF583700</td>\n",
       "      <td>82</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>50099104</td>\n",
       "      <td>50099353</td>\n",
       "      <td>MLLVQGQHQNEEGLTRHLLSSSFTLSLPTPSFPLPHKVPMCLYPPL...</td>\n",
       "      <td>ATGCTGTTGGTTCAAGGACAACACCAGAATGAAGAGGGTCTCACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_999</th>\n",
       "      <td>31</td>\n",
       "      <td>HF548109</td>\n",
       "      <td>60</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2718365</td>\n",
       "      <td>2722632</td>\n",
       "      <td>MKVMQTLMQEDHYHLKQLSFLLKRVSQRNFHLVFWIFLFVLEFHLI...</td>\n",
       "      <td>ATGAAAGTAATGCAGACACTTATGCAGGAAGACCACTACCATCTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21483 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                PS                  Description  Plength     Gap_Gene?    Chr  \\\n",
       "Name                                                                            \n",
       "Baz_Hs_1        31  ENST00000308604.5_18272_457       52  Not_Gap_Gene   chr2   \n",
       "Baz_Hs_10       31  ENST00000411630.2_23991_594       59  Not_Gap_Gene   chr4   \n",
       "Baz_Hs_103      31  ENST00000499346.2_27384_333       84  Not_Gap_Gene   chr5   \n",
       "Baz_Hs_108      24  ENST00000501177.3_12701_390       84  Not_Gap_Gene  chr16   \n",
       "Baz_Hs_112      25  ENST00000503704.1_24073_293       52  Not_Gap_Gene   chr4   \n",
       "...             ..                          ...      ...           ...    ...   \n",
       "vdp2013_S4_994  31                     HF584391       69  Not_Gap_Gene   chrX   \n",
       "vdp2013_S4_995  31                     HF583960       46  Not_Gap_Gene  chr10   \n",
       "vdp2013_S4_997  31                     HF548108       40  Not_Gap_Gene   chr6   \n",
       "vdp2013_S4_998  31                     HF583700       82  Not_Gap_Gene  chr12   \n",
       "vdp2013_S4_999  31                     HF548109       60  Not_Gap_Gene  chr18   \n",
       "\n",
       "                Gene_Start_hg38  Gene_End_hg38 Strand  CDS_Start_hg38  \\\n",
       "Name                                                                    \n",
       "Baz_Hs_1                    NaN            NaN      -       111429413   \n",
       "Baz_Hs_10                   NaN            NaN      +        52713673   \n",
       "Baz_Hs_103                  NaN            NaN      -       128082767   \n",
       "Baz_Hs_108                  NaN            NaN      -        54919086   \n",
       "Baz_Hs_112                  NaN            NaN      -        82900166   \n",
       "...                         ...            ...    ...             ...   \n",
       "vdp2013_S4_994              NaN            NaN      -       143628143   \n",
       "vdp2013_S4_995              NaN            NaN      +       104020518   \n",
       "vdp2013_S4_997              NaN            NaN      +        70860637   \n",
       "vdp2013_S4_998              NaN            NaN      +        50099104   \n",
       "vdp2013_S4_999              NaN            NaN      +         2718365   \n",
       "\n",
       "                CDS_End_hg38  \\\n",
       "Name                           \n",
       "Baz_Hs_1           111429572   \n",
       "Baz_Hs_10           52713853   \n",
       "Baz_Hs_103         128083022   \n",
       "Baz_Hs_108          54925136   \n",
       "Baz_Hs_112          82900435   \n",
       "...                      ...   \n",
       "vdp2013_S4_994     143628353   \n",
       "vdp2013_S4_995     104021665   \n",
       "vdp2013_S4_997      70860760   \n",
       "vdp2013_S4_998      50099353   \n",
       "vdp2013_S4_999       2722632   \n",
       "\n",
       "                                                 Protein_Sequence  \\\n",
       "Name                                                                \n",
       "Baz_Hs_1        MTDTENHDSSPSSTSTCCPPITAGMQLKDSLGPGSNCPLWTLRPLH...   \n",
       "Baz_Hs_10       MLVATGQCSRCFMFTFSTFSFNCHNSEVDSVRDRLPQDHSAPANSM...   \n",
       "Baz_Hs_103      MLGAFRSGPQPLPEPRARCVPQPGLLWALTRRRESPLVTPGLNLEE...   \n",
       "Baz_Hs_108      MLAEIHPKAGLQSLQFIMELLYWLLEGGDSEDKEDATGNVEMKNIQ...   \n",
       "Baz_Hs_112      MRSREAGPKLRRIQEPANGSPGAVSETGGYREERLSDAEIMGKLLA...   \n",
       "...                                                           ...   \n",
       "vdp2013_S4_994  MLYTHNTEFNLKRQICFVPQCKTFVSLCFVKQTQENWYTCTSWVLY...   \n",
       "vdp2013_S4_995     MREWLSIRNMRIKCEIFSCSVKPMSANCISCRMKNATCWLSMRLRN   \n",
       "vdp2013_S4_997           MFAYKGSSYHVSNTSNSINPTPKLASNPVGRYCMIKCLII   \n",
       "vdp2013_S4_998  MLLVQGQHQNEEGLTRHLLSSSFTLSLPTPSFPLPHKVPMCLYPPL...   \n",
       "vdp2013_S4_999  MKVMQTLMQEDHYHLKQLSFLLKRVSQRNFHLVFWIFLFVLEFHLI...   \n",
       "\n",
       "                                                     CDS_Sequence  \n",
       "Name                                                               \n",
       "Baz_Hs_1        ATGACAGACACTGAAAATCACGACTCATCCCCCTCCAGCACCTCTA...  \n",
       "Baz_Hs_10       ATGCTGGTGGCAACAGGGCAGTGTAGCAGGTGCTTCATGTTCACCT...  \n",
       "Baz_Hs_103      ATGCTGGGGGCTTTCCGGTCGGGGCCGCAGCCGCTTCCGGAGCCGC...  \n",
       "Baz_Hs_108      ATGTTGGCTGAAATTCATCCCAAGGCTGGTCTGCAAAGTCTGCAAT...  \n",
       "Baz_Hs_112      ATGCGAAGCAGAGAGGCAGGACCAAAATTGAGGCGAATCCAGGAAC...  \n",
       "...                                                           ...  \n",
       "vdp2013_S4_994  ATGCTTTATACACATAATACTGAATTTAACCTCAAGAGGCAAATCT...  \n",
       "vdp2013_S4_995  ATGAGAGAATGGCTCAGCATCAGAAACATGAGAATCAAATGCGAGA...  \n",
       "vdp2013_S4_997  ATGTTTGCATATAAGGGAAGTAGTTATCATGTTAGTAATACCTCTA...  \n",
       "vdp2013_S4_998  ATGCTGTTGGTTCAAGGACAACACCAGAATGAAGAGGGTCTCACAA...  \n",
       "vdp2013_S4_999  ATGAAAGTAATGCAGACACTTATGCAGGAAGACCACTACCATCTAA...  \n",
       "\n",
       "[21483 rows x 12 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Internal duplicates taken care of, now time to do the external (AG and UG)\n",
    "# Remove controls which overlap with known annotated or unannotated genes \n",
    "\n",
    "## now reading in the phylo_df \n",
    "phylo_df = pd.read_csv(r'gs://ug-wphu/gtex_analysis/victor_2149+Ens89/Hs_Ens89+2149_PS_seq_etc_hg38.txt',sep='\\t')\n",
    "phylo_df.rename(columns={'Name':'Gene_ID'},inplace=True)\n",
    "phylo_df.rename(columns={'Protein_ID':'Name'},inplace=True)\n",
    "phylo_df.set_index('Name',inplace=True)\n",
    "phylo_df.drop('Gene_ID',inplace=True,axis=1)\n",
    "phylo_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e65e71ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill in Gene Start if it is undefined, so we can filter from one set of columns\n",
    "for xi,yi in phylo_df.iterrows():\n",
    "    if str(yi['Gene_Start_hg38'] == 'nan'):\n",
    "        phylo_df.loc[xi,'Gene_Start_hg38'] = phylo_df.loc[xi,'CDS_Start_hg38']\n",
    "        phylo_df.loc[xi,'Gene_End_hg38'] = phylo_df.loc[xi,'CDS_End_hg38']\n",
    "    if 'ENSP' in xi[:4]:\n",
    "        # print('annotated')\n",
    "        phylo_df.loc[xi,'Status'] = 'annotated'\n",
    "    else:\n",
    "        phylo_df.loc[xi,'Status'] = 'unannotated'\n",
    "    phylo_df.loc[xi,'Control_Set'] = 'victorgenes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5db34c2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr14_GL000009v2_random\n",
      "chr15\n",
      "chr15_KI270850v1_alt\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr22_KI270879v1_alt\n",
      "chr3\n",
      "chr4\n",
      "chr4_GL000008v2_random\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr7_KI270803v1_alt\n",
      "chr8\n",
      "chr8_KI270821v1_alt\n",
      "chr9\n",
      "chrUn_KI270742v1\n",
      "chrX\n",
      "chrY\n",
      "Total ORF/NORFs removed due to overlap with victor genes:  480\n",
      "For UG:  182\n",
      "For Ensembl:  298\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(38888, 10)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Code to: \n",
    "# remove overlapping controls, very similar to code above \n",
    "ug_cleaned_dfs = []\n",
    "ugag_removal_list = []\n",
    "\n",
    "victor_overlap = 0\n",
    "ug_overlap = 0\n",
    "ag_overlap = 0\n",
    "\n",
    "for unique_chrom in sorted(fixed_df.Chr.unique()):\n",
    "    print(unique_chrom)\n",
    "    by_chrom_df = fixed_df[fixed_df.Chr==unique_chrom]\n",
    "    by_chrom_df_dd = by_chrom_df.sort_values(by=[gi,ge])\n",
    "    \n",
    "    phylo_by_chr = phylo_df[phylo_df.Chr==unique_chrom].sort_values(by=[gi,ge])\n",
    "    \n",
    "    total_dups += dupls.shape[0]\n",
    "    \n",
    "    for xi,yi in by_chrom_df_dd.iterrows():\n",
    "        \n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "        \n",
    "        query_new = phylo_by_chr[(phylo_by_chr[gi]>(index_start-1000000))&(phylo_by_chr[ge]<(index_end+1000000))]\n",
    "        \n",
    "        for xii,yii in query_new.iterrows():\n",
    "            \n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "            \n",
    "            case_1 = query_start in range(index_start,index_end+1)\n",
    "            case_2 = query_end in range(index_start,index_end+1)\n",
    "            case_3 = index_start in range(int(query_start),int(query_end)+1)\n",
    "            \n",
    "            if (xi!=xii):\n",
    "                if any([case_1,case_2,case_3]):\n",
    "                    by_chrom_df_dd.loc[xi,'removal']=True\n",
    "                    victor_overlap += 1\n",
    "                    \n",
    "                    if yii['Status']=='annotated':\n",
    "                        ag_overlap += 1\n",
    "                    elif yii['Status']=='unannotated':\n",
    "                        ug_overlap += 1\n",
    "                    else:\n",
    "                        raise Exception('ayylmao')\n",
    "\n",
    "     \n",
    "    ugag_removal_list.append(by_chrom_df_dd[by_chrom_df_dd.removal])\n",
    "    by_chrom_df_dd = by_chrom_df_dd[~by_chrom_df_dd.removal]\n",
    "    ug_cleaned_dfs.append(by_chrom_df_dd)\n",
    "    \n",
    "print('Total ORF/NORFs removed due to overlap with victor genes: ',victor_overlap)\n",
    "print('For UG: ',ug_overlap)\n",
    "print('For Ensembl: ',ag_overlap)\n",
    "\n",
    "df2 = pd.concat(ug_cleaned_dfs)\n",
    "\n",
    "df2.shape  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "158ed2e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr14_GL000009v2_random\n",
      "chr15\n",
      "chr15_KI270850v1_alt\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr22_KI270879v1_alt\n",
      "chr3\n",
      "chr4\n",
      "chr4_GL000008v2_random\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr7_KI270803v1_alt\n",
      "chr8\n",
      "chr8_KI270821v1_alt\n",
      "chr9\n",
      "chrUn_KI270742v1\n",
      "chrX\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "# Checking again, with Exceptions, to ensure we did not miss anything obvious \n",
    "\n",
    "for chrom_id in df2.Chr.unique():\n",
    "\n",
    "    print(chrom_id)\n",
    "\n",
    "    query_df = df2[df2['Chr']==chrom_id]\n",
    "\n",
    "    for xi,yi in query_df.iterrows():\n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "        \n",
    "        p_by_c = phylo_df[phylo_df.Chr==chrom_id]\n",
    "        query_new = p_by_c[(p_by_c[gi]>(index_start-1000000))&(p_by_c[ge]<(index_end+1000000))]\n",
    "\n",
    "        for xii,yii in query_new.iterrows():\n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "\n",
    "            if xi!=xii:\n",
    "                if query_start in range(index_start,index_end+1) or query_end in range(index_start,index_end+1):\n",
    "                    print(xi,yi)\n",
    "                    print(xii,yii) \n",
    "                    raise Exception('DANGER')\n",
    "\n",
    "                elif index_start in range(int(query_start),int(query_end)+1):\n",
    "                    raise Exception('DANGER')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ba2a7fc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chr                   chr17\n",
       "Gene_Start_hg38    41936883\n",
       "Gene_End_hg38      41937015\n",
       "Inclusivity               1\n",
       "Strand                    -\n",
       "Control_Set               3\n",
       "Status                  orf\n",
       "user_length             132\n",
       "removal                True\n",
       "heirarchy                 3\n",
       "Name: GRCh37.Ens87.dna_rm.chr17.intergenic_gt122.orf:40093136-40093267:-orf_5_control_set_3, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a sequence that should've been removed \n",
    "ugag_removal_list[len(ugag_removal_list)//3].iloc[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e0d7022c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chr, Gene_Start_hg38, Gene_End_hg38, Inclusivity, Strand, Control_Set, Status, user_length, removal, heirarchy]\n",
       "Index: []"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# making sure that it's actually removed \n",
    "df2[(df2.Chr=='chr17')&(df2.Gene_Start_hg38==41936883)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ea91a6cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Inclusivity</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504-43241.norf_segment:42671-42895:+norf_1_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42671</td>\n",
       "      <td>42896</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>225</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:58253-58618:-norf_1_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>58253</td>\n",
       "      <td>58619</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>366</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208-60377.norf_segment:59878-60066:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59878</td>\n",
       "      <td>60067</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>189</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:105742-106428.norf_segment:105843-105989:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>105843</td>\n",
       "      <td>105990</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>norf</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:106531-107479.norf_segment:107280-107417:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107280</td>\n",
       "      <td>107418</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>138</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59027971-59028150:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>180</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59209645-59209812:+orf_9_control_set_5</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57063496</td>\n",
       "      <td>57063664</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "      <td>168</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59339711-59339869:+orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57193560</td>\n",
       "      <td>57193719</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>159</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59353081-59353329:+orf_5_control_set_3</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57206930</td>\n",
       "      <td>57207179</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>orf</td>\n",
       "      <td>249</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:59355200-59355355:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>156</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...  chr1            42671   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...  chr1            58253   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...  chr1            59878   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...  chr1           105843   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...  chr1           107280   \n",
       "...                                                  ...              ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57063496   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57193560   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57206930   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  \\\n",
       "Name                                                                \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...          42896   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...          58619   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...          60067   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...         105990   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...         107418   \n",
       "...                                                           ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       56882004   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57063664   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57193719   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57207179   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...       57209205   \n",
       "\n",
       "                                                    Inclusivity Strand  \\\n",
       "Name                                                                     \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...            1      +   \n",
       "...                                                         ...    ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      -   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      +   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...            1      -   \n",
       "\n",
       "                                                   Control_Set Status  \\\n",
       "Name                                                                    \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...           1   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...           5   norf   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...           2   norf   \n",
       "...                                                        ...    ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           1    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           5    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           1    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           3    orf   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...           4    orf   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...          225    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...          366    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...          189    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...          147    False   \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...          138    False   \n",
       "...                                                         ...      ...   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          180    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          168    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          159    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          249    False   \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          156    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:42504...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:55608...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:59208...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10574...          4  \n",
       "GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:10653...          4  \n",
       "...                                                       ...  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:5...          3  \n",
       "\n",
       "[38888 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Display the final dataframe to be processed further, and create dictionary for separate dataframes\n",
    "display(df2)\n",
    "df_by_set = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "faab531b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to turn the DFs as we have them formatted into properly formatted BEDs with the control set in their name\n",
    "def bedify_2(input_df):\n",
    "    #display(input_df)\n",
    "    ret_df = input_df.reset_index()\n",
    "    newnames = ret_df.Name#+ret_df.loc[xi,'Status']+'_control_set_'+str(ret_df.loc[xi,'Control_Set']) for xi in ret_df.index] \n",
    "    ret_df['new_index'] = newnames\n",
    "    ret_df_2 = ret_df[['Chr','Gene_Start_hg38','Gene_End_hg38','new_index','Strand']]\n",
    "    ret_df_2['zeroes'] = 0 # THIS IS ONLY BECUASE WE FIX FOR INCLUSIVITY BEFOREHAND \n",
    "    ret_df_2 = ret_df_2[['Chr','Gene_Start_hg38','Gene_End_hg38','new_index','zeroes','Strand']].sort_values(by=['Chr','Gene_Start_hg38','Gene_End_hg38'])\n",
    "    ret_df_2.new_index[1]\n",
    "    return ret_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "aa318677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14',\n",
       "       'chr14_GL000009v2_random', 'chr15', 'chr15_KI270850v1_alt',\n",
       "       'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21',\n",
       "       'chr22', 'chr22_KI270879v1_alt', 'chr3', 'chr4',\n",
       "       'chr4_GL000008v2_random', 'chr5', 'chr6', 'chr7',\n",
       "       'chr7_KI270803v1_alt', 'chr8', 'chr8_KI270821v1_alt', 'chr9',\n",
       "       'chrUn_KI270742v1', 'chrX', 'chrY'], dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check for unique chrs - we do INDEED have non-standard contigs\n",
    "df2.Chr.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "db8e7cee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "math check: 3926 + 3897 == 7823 ? True \n",
      "math check: 3908 + 3872 == 7780 ? True \n",
      "math check: 3903 + 3876 == 7779 ? True \n",
      "math check: 3881 + 3884 == 7765 ? True \n",
      "math check: 3863 + 3878 == 7741 ? True \n"
     ]
    }
   ],
   "source": [
    "# Output separate beds for each type for each control set\n",
    "\n",
    "for control_set in sorted(df2.Control_Set.unique()):\n",
    "    # print(control_set)\n",
    "    df_temp = df2[df2.Control_Set == control_set].sort_values(by=['Chr','Gene_Start_hg38','Gene_End_hg38'])\n",
    "    df_temp_orf = df_temp[~df_temp.Status.str.contains('norf')]\n",
    "    df_temp_norf = df_temp[df_temp.Status.str.contains('norf')]\n",
    "    print(f'math check: {df_temp_orf.shape[0]} + {df_temp_norf.shape[0]} == {df_temp.shape[0]} ? {df_temp_orf.shape[0] + df_temp_norf.shape[0] == df_temp.shape[0]} ')\n",
    "    df_by_set[str(control_set)] = {'orf':bedify_2(df_temp_orf),'norf':bedify_2(df_temp_norf)}\n",
    "    bedify_2(df_temp_orf).to_csv(f'orf_{control_set}_v5.bam',sep='\\t',index=False,header=False,encoding=\"utf-8\")\n",
    "    bedify_2(df_temp_norf).to_csv(f'norf_{control_set}_v5.bam',sep='\\t',index=False,header=False,encoding=\"utf-8\")\n",
    "    # it's going to be some work to convert these back to GRCh37 - key the original by the names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6db9774d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>new_index</th>\n",
       "      <th>zeroes</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>478725</td>\n",
       "      <td>478917</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>675222</td>\n",
       "      <td>675432</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2313025</td>\n",
       "      <td>2313796</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2921177</td>\n",
       "      <td>2921342</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3019495</td>\n",
       "      <td>3019666</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24152772</td>\n",
       "      <td>24152943</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25409413</td>\n",
       "      <td>25409626</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25953496</td>\n",
       "      <td>25953709</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26530763</td>\n",
       "      <td>26530898</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57063496</td>\n",
       "      <td>57063664</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3863 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chr  Gene_Start_hg38  Gene_End_hg38  \\\n",
       "0     chr1           478725         478917   \n",
       "1     chr1           675222         675432   \n",
       "2     chr1          2313025        2313796   \n",
       "3     chr1          2921177        2921342   \n",
       "4     chr1          3019495        3019666   \n",
       "...    ...              ...            ...   \n",
       "3858  chrY         24152772       24152943   \n",
       "3859  chrY         25409413       25409626   \n",
       "3860  chrY         25953496       25953709   \n",
       "3861  chrY         26530763       26530898   \n",
       "3862  chrY         57063496       57063664   \n",
       "\n",
       "                                              new_index  zeroes Strand  \n",
       "0     GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...       0      +  \n",
       "1     GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...       0      -  \n",
       "2     GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...       0      -  \n",
       "3     GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...       0      +  \n",
       "4     GRCh37.Ens87.dna_rm.chr1.intergenic_gt122.orf:...       0      +  \n",
       "...                                                 ...     ...    ...  \n",
       "3858  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "3859  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "3860  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "3861  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "3862  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "\n",
       "[3863 rows x 6 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bedify_2(df_temp_orf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "87b06a1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>new_index</th>\n",
       "      <th>zeroes</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42671</td>\n",
       "      <td>42896</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>58253</td>\n",
       "      <td>58619</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59878</td>\n",
       "      <td>60067</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>105843</td>\n",
       "      <td>105990</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107280</td>\n",
       "      <td>107418</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38883</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38884</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57063496</td>\n",
       "      <td>57063664</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38885</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57193560</td>\n",
       "      <td>57193719</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38886</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57206930</td>\n",
       "      <td>57207179</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38887</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Gene_Start_hg38  Gene_End_hg38  \\\n",
       "0      chr1            42671          42896   \n",
       "1      chr1            58253          58619   \n",
       "2      chr1            59878          60067   \n",
       "3      chr1           105843         105990   \n",
       "4      chr1           107280         107418   \n",
       "...     ...              ...            ...   \n",
       "38883  chrY         56881824       56882004   \n",
       "38884  chrY         57063496       57063664   \n",
       "38885  chrY         57193560       57193719   \n",
       "38886  chrY         57206930       57207179   \n",
       "38887  chrY         57209049       57209205   \n",
       "\n",
       "                                               new_index  zeroes Strand  \n",
       "0      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...       0      +  \n",
       "1      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...       0      -  \n",
       "2      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...       0      -  \n",
       "3      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...       0      +  \n",
       "4      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...       0      +  \n",
       "...                                                  ...     ...    ...  \n",
       "38883  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "38884  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38885  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38886  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38887  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "\n",
       "[38888 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bedify_2(df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a3c9c888",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>new_index</th>\n",
       "      <th>zeroes</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42671</td>\n",
       "      <td>42896</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>58253</td>\n",
       "      <td>58619</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59878</td>\n",
       "      <td>60067</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>105843</td>\n",
       "      <td>105990</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107280</td>\n",
       "      <td>107418</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38883</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38884</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57063496</td>\n",
       "      <td>57063664</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38885</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57193560</td>\n",
       "      <td>57193719</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38886</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57206930</td>\n",
       "      <td>57207179</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38887</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Gene_Start_hg38  Gene_End_hg38  \\\n",
       "0      chr1            42671          42896   \n",
       "1      chr1            58253          58619   \n",
       "2      chr1            59878          60067   \n",
       "3      chr1           105843         105990   \n",
       "4      chr1           107280         107418   \n",
       "...     ...              ...            ...   \n",
       "38883  chrY         56881824       56882004   \n",
       "38884  chrY         57063496       57063664   \n",
       "38885  chrY         57193560       57193719   \n",
       "38886  chrY         57206930       57207179   \n",
       "38887  chrY         57209049       57209205   \n",
       "\n",
       "                                               new_index  zeroes Strand  \n",
       "0      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...       0      +  \n",
       "1      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...       0      -  \n",
       "2      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...       0      -  \n",
       "3      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...       0      +  \n",
       "4      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...       0      +  \n",
       "...                                                  ...     ...    ...  \n",
       "38883  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "38884  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38885  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38886  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38887  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "\n",
       "[38888 rows x 6 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_final = bedify_2(df2) # FINAL of complete GRCh37->GRCh38 controls \n",
    "grch37_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "40b6a258",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write this out , later to be passed through GTEx and RNASeq calling and assured that expression is as expected\n",
    "grch37_final.to_csv('grch37liftover_38888.bed',sep='\\t',index=False,header=False,encoding=\"utf-8\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "40d693bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>new_index</th>\n",
       "      <th>zeroes</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42671</td>\n",
       "      <td>42896</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>58253</td>\n",
       "      <td>58619</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59878</td>\n",
       "      <td>60067</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>105843</td>\n",
       "      <td>105990</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107280</td>\n",
       "      <td>107418</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38883</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38884</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57063496</td>\n",
       "      <td>57063664</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38885</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57193560</td>\n",
       "      <td>57193719</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38886</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57206930</td>\n",
       "      <td>57207179</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38887</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Gene_Start_hg38  Gene_End_hg38  \\\n",
       "0      chr1            42671          42896   \n",
       "1      chr1            58253          58619   \n",
       "2      chr1            59878          60067   \n",
       "3      chr1           105843         105990   \n",
       "4      chr1           107280         107418   \n",
       "...     ...              ...            ...   \n",
       "38883  chrY         56881824       56882004   \n",
       "38884  chrY         57063496       57063664   \n",
       "38885  chrY         57193560       57193719   \n",
       "38886  chrY         57206930       57207179   \n",
       "38887  chrY         57209049       57209205   \n",
       "\n",
       "                                               new_index  zeroes Strand  \n",
       "0      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:4250...       0      +  \n",
       "1      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5560...       0      -  \n",
       "2      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:5920...       0      -  \n",
       "3      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1057...       0      +  \n",
       "4      GRCh37.Ens87.dna_rm.chr1.intergenic_gt122:1065...       0      +  \n",
       "...                                                  ...     ...    ...  \n",
       "38883  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "38884  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38885  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38886  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      +  \n",
       "38887  GRCh37.Ens87.dna_rm.chrY.intergenic_gt122.orf:...       0      -  \n",
       "\n",
       "[38888 rows x 6 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1b06be5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e948857",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
