{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f94d8e26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "# Daniel Marten\n",
    "# Custom notebook to remove controls that don't map cleanly to GRCh37\n",
    "\n",
    "import pandas as pd\n",
    "import qtl.io as io\n",
    "import qtl.norm as norm\n",
    "import csv\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "92f9ead5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39172</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39173</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26627091</td>\n",
       "      <td>26627364</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39174</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39175</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57070959</td>\n",
       "      <td>57071091</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39176</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39177 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2  \\\n",
       "0      chr1     35726     35855   \n",
       "1      chr1     57850     58015   \n",
       "2      chr1     59864     60014   \n",
       "3      chr1     67741     67906   \n",
       "4      chr1    108819    108960   \n",
       "...     ...       ...       ...   \n",
       "39172  chrY  26623039  26623183   \n",
       "39173  chrY  26627091  26627364   \n",
       "39174  chrY  56881824  56882004   \n",
       "39175  chrY  57070959  57071091   \n",
       "39176  chrY  57209049  57209205   \n",
       "\n",
       "                                                       3  4  5  \n",
       "0      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...  0  -  \n",
       "1      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...  0  -  \n",
       "2      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...  0  -  \n",
       "3      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...  0  +  \n",
       "4      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...  0  +  \n",
       "...                                                  ... .. ..  \n",
       "39172  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  0  -  \n",
       "39173  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  0  -  \n",
       "39174  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  0  -  \n",
       "39175  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...  0  -  \n",
       "39176  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  0  -  \n",
       "\n",
       "[39177 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Version of the BED file read in \n",
    "bed_001 = pd.read_csv(r'grch38_bed_complete_allsets.bed',sep='\\t',header=None)\n",
    "bed_001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "41780582",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(bed_001.duplicated([0,1])) # no duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cef4ac59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39022</th>\n",
       "      <td>chrY</td>\n",
       "      <td>28769186</td>\n",
       "      <td>28769330</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39023</th>\n",
       "      <td>chrY</td>\n",
       "      <td>28773238</td>\n",
       "      <td>28773511</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39024</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59027971</td>\n",
       "      <td>59028151</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39025</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59217108</td>\n",
       "      <td>59217240</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39026</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59355200</td>\n",
       "      <td>59355356</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39027 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2  \\\n",
       "0      chr1     35726     35855   \n",
       "1      chr1     57850     58015   \n",
       "2      chr1     59864     60014   \n",
       "3      chr1     67741     67906   \n",
       "4      chr1    108819    108960   \n",
       "...     ...       ...       ...   \n",
       "39022  chrY  28769186  28769330   \n",
       "39023  chrY  28773238  28773511   \n",
       "39024  chrY  59027971  59028151   \n",
       "39025  chrY  59217108  59217240   \n",
       "39026  chrY  59355200  59355356   \n",
       "\n",
       "                                                       3  4  5  \n",
       "0      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...  1  -  \n",
       "1      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...  1  -  \n",
       "2      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...  1  -  \n",
       "3      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...  1  +  \n",
       "4      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...  1  +  \n",
       "...                                                  ... .. ..  \n",
       "39022  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  1  -  \n",
       "39023  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  1  -  \n",
       "39024  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  1  -  \n",
       "39025  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...  1  -  \n",
       "39026  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...  1  -  \n",
       "\n",
       "[39027 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in result of listing our complete GRCh38 BED to GRCh37 with default settings\n",
    "# with: https://genome.ucsc.edu/cgi-bin/hgLiftOver\n",
    "bed_to_keep = pd.read_csv(r'/Users/marten/Downloads/hglft_genome_32ff1_a88a10.bed',sep='\\t',header=None)\n",
    "bed_to_keep "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fed030e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(bed_to_keep.duplicated([0,1])) # three existing controls were lifted over to same location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "193d700c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Proper renaming\n",
    "bed_001.rename(columns={0:'Chr',1:'Gene_Start_hg38',2:'Gene_End_hg38',3:'Name',4:'Quality',5:'Strand'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1ad5580a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Further renaming for GRCh37\n",
    "bed_to_keep.rename(columns={0:'Chr',1:'Gene_Start_GRCh37_Lift',2:'Gene_End_GRCh37_Lift',3:'Name',4:'Quality',5:'Strand'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b55bb5b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_unique = bed_to_keep.Chr.unique() # unique chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8b7f79ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr2\n",
      "chr5\n",
      "chr19\n",
      "chr16\n",
      "chr9\n",
      "chr1_gl000192_random\n",
      "dropped!\n",
      "chr10\n",
      "chrUn_gl000228\n",
      "dropped!\n",
      "chr11\n",
      "chr11_gl000202_random\n",
      "dropped!\n",
      "chr12\n",
      "chrUn_gl000212\n",
      "dropped!\n",
      "chr21\n",
      "chr13\n",
      "chrUn_gl000234\n",
      "dropped!\n",
      "chr14\n",
      "chr15\n",
      "chr17\n",
      "chr17_gl000204_random\n",
      "dropped!\n",
      "chr18\n",
      "chr20\n",
      "chrY\n",
      "chr7_gl000195_random\n",
      "dropped!\n",
      "chrUn_gl000220\n",
      "dropped!\n",
      "chr17_gl000205_random\n",
      "dropped!\n",
      "chr4_gl000193_random\n",
      "dropped!\n",
      "chrUn_gl000217\n",
      "dropped!\n",
      "chrUn_gl000244\n",
      "dropped!\n",
      "chrUn_gl000241\n",
      "dropped!\n",
      "chrUn_gl000237\n",
      "dropped!\n",
      "chrUn_gl000233\n",
      "dropped!\n",
      "chrUn_gl000232\n",
      "dropped!\n",
      "chrUn_gl000229\n",
      "dropped!\n",
      "chrUn_gl000224\n",
      "dropped!\n",
      "chr4\n",
      "chrUn_gl000240\n",
      "dropped!\n",
      "chrUn_gl000235\n",
      "dropped!\n",
      "chr22\n",
      "chr3\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chrUn_gl000211\n",
      "dropped!\n",
      "chrX\n"
     ]
    }
   ],
   "source": [
    "# Removing NONCANONICAL Chromosome controls, since they don't map cleanly\n",
    "chr_cannon = []\n",
    "for chri in chr_unique:\n",
    "    print(chri)\n",
    "    if len(chri)<7:\n",
    "        chr_cannon.append(chri)\n",
    "    else:\n",
    "        print('dropped!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c3b1afe9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr1',\n",
       " 'chr2',\n",
       " 'chr5',\n",
       " 'chr19',\n",
       " 'chr16',\n",
       " 'chr9',\n",
       " 'chr10',\n",
       " 'chr11',\n",
       " 'chr12',\n",
       " 'chr21',\n",
       " 'chr13',\n",
       " 'chr14',\n",
       " 'chr15',\n",
       " 'chr17',\n",
       " 'chr18',\n",
       " 'chr20',\n",
       " 'chrY',\n",
       " 'chr4',\n",
       " 'chr22',\n",
       " 'chr3',\n",
       " 'chr6',\n",
       " 'chr7',\n",
       " 'chr8',\n",
       " 'chrX']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_cannon # list of canonical chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4b5e66fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chr_cannon) # 22 and two sex chrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "13fbac96",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_to_keep = bed_to_keep[bed_to_keep.Chr.isin(chr_cannon)] # only keep ones which map cleanly to canonical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5e6d80fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39022</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39023</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39024</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39025</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39026</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38948 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Name\n",
       "0      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...\n",
       "1      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...\n",
       "2      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...\n",
       "3      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...\n",
       "4      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...\n",
       "...                                                  ...\n",
       "39022  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "39023  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "39024  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "39025  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...\n",
       "39026  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "\n",
       "[38948 rows x 1 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Miniature check of their names\n",
    "bed_to_keep_mini = bed_to_keep[['Name']]\n",
    "bed_to_keep_mini"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4fc9bb31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To remove duplicates - only keep the first instance of each name \n",
    "bed_to_keep_mini = bed_to_keep_mini[~bed_to_keep_mini.duplicated(['Name'],keep='first')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ae482a52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39022</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39023</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39024</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39025</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39026</th>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Name\n",
       "0      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...\n",
       "1      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:5560...\n",
       "2      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...\n",
       "3      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:6663...\n",
       "4      GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:1086...\n",
       "...                                                  ...\n",
       "39022  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "39023  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "39024  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "39025  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:5707...\n",
       "39026  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...\n",
       "\n",
       "[38888 rows x 1 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_to_keep_mini"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1155f26d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedo = bed_001.set_index('Name').join(bed_to_keep_mini.set_index('Name'),how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "983f317e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-orf_6_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:57850-58014:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:59864-60013:-orf_8_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632-67997.norf_segment:67741-67905:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:108680-109563.norf_segment:108819-108959:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26623039-26623182:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26627091-26627363:-orf_1_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26627091</td>\n",
       "      <td>26627364</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:56881824-56882003:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070946-57071187.norf_segment:57070959-57071090:-norf_2_control_set_2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57070959</td>\n",
       "      <td>57071091</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:57209049-57209204:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1            35726   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  chr1            57850   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  chr1            59864   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  chr1            67741   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  chr1           108819   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26623039   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26627091   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  chrY         57070959   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  Quality  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          35855        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...          58015        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...          60014        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...          67906        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...         108960        0   \n",
       "...                                                           ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26623183        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26627364        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       56882004        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       57071091        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       57209205        0   \n",
       "\n",
       "                                                   Strand  \n",
       "Name                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      -  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...      -  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...      -  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...      +  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...      +  \n",
       "...                                                   ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...      -  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -  \n",
       "\n",
       "[38888 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bedo.dropna() # visual of table with no NaN values "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7fcda9b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking no duplicate start sites \n",
    "sum(bedo.duplicated(['Chr','Gene_Start_hg38']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f40423da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({2: 44, 3: 5, 7: 1})"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## TESTING\n",
    "from collections import Counter\n",
    "\n",
    "Counter(list(bed_to_keep[bed_to_keep.duplicated(['Name'],keep=False)].Name.value_counts()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8aabf588",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedo['Length'] = bedo.Gene_End_hg38 - bedo.Gene_Start_hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d204601e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-orf_6_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:57850-58014:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:59864-60013:-orf_8_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632-67997.norf_segment:67741-67905:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:108680-109563.norf_segment:108819-108959:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26623039-26623182:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26627091-26627363:-orf_1_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26627091</td>\n",
       "      <td>26627364</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:56881824-56882003:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070946-57071187.norf_segment:57070959-57071090:-norf_2_control_set_2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57070959</td>\n",
       "      <td>57071091</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:57209049-57209204:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1            35726   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  chr1            57850   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  chr1            59864   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  chr1            67741   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  chr1           108819   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26623039   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26627091   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  chrY         57070959   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  Quality  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          35855        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...          58015        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...          60014        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...          67906        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...         108960        0   \n",
       "...                                                           ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26623183        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26627364        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       56882004        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       57071091        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       57209205        0   \n",
       "\n",
       "                                                   Strand  Length  \n",
       "Name                                                               \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      -     129  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...      -     165  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...      -     150  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...      +     165  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...      +     141  \n",
       "...                                                   ...     ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     144  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     273  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     180  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...      -     132  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     156  \n",
       "\n",
       "[38888 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bedo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9d5d4c15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to turn the DFs as we have them formatted into properly formatted BEDs with the control set in their name\n",
    "def bedify(input_df):\n",
    "    ret_df = input_df.reset_index()\n",
    "    #newnames = [ret_df.loc[xi,'Name']+ret_df.loc[xi,'Status']+'_control_set_'+str(ret_df.loc[xi,'Control_Set']) for xi in ret_df.index] \n",
    "    ret_df['new_index'] = ret_df.Name # already included at this point in pipeline\n",
    "    ret_df_2 = ret_df[['Chr','Gene_Start_hg38','Gene_End_hg38','new_index','Strand']]\n",
    "    ret_df_2['zeroes'] = 0\n",
    "    ret_df_2 = ret_df_2[['Chr','Gene_Start_hg38','Gene_End_hg38','new_index','zeroes','Strand']]\n",
    "    ret_df_2.new_index[1]\n",
    "    return ret_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ec1fb8e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedo['Control_Set'] = [xi[-1] for xi in bedo.index.to_series()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "86af649b",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedo['Status'] = [xi.split(':')[-1].split('_')[0][1:] for xi in bedo.index.to_series()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "91d99da5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_by_set = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9790e16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-orf_6_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>129</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:57850-58014:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:59864-60013:-orf_8_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>150</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632-67997.norf_segment:67741-67905:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:108680-109563.norf_segment:108819-108959:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>141</td>\n",
       "      <td>5</td>\n",
       "      <td>norf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26623039-26623182:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26627091-26627363:-orf_1_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26627091</td>\n",
       "      <td>26627364</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:56881824-56882003:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070946-57071187.norf_segment:57070959-57071090:-norf_2_control_set_2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57070959</td>\n",
       "      <td>57071091</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>132</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:57209049-57209204:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>156</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1            35726   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  chr1            57850   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  chr1            59864   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  chr1            67741   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  chr1           108819   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26623039   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26627091   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  chrY         57070959   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  Quality  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          35855        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...          58015        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...          60014        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...          67906        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...         108960        0   \n",
       "...                                                           ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26623183        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26627364        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       56882004        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       57071091        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       57209205        0   \n",
       "\n",
       "                                                   Strand  Length Control_Set  \\\n",
       "Name                                                                            \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      -     129           4   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...      -     165           1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...      -     150           5   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...      +     165           2   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...      +     141           5   \n",
       "...                                                   ...     ...         ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     144           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     273           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     180           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...      -     132           2   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     156           4   \n",
       "\n",
       "                                                   Status  \n",
       "Name                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...    orf  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...   norf  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...    orf  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...   norf  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...   norf  \n",
       "...                                                   ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...    orf  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...   norf  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...    orf  \n",
       "\n",
       "[38888 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bedo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1b320950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "quick check: 3916 + 3889 == 7805 ? True \n",
      "quick check: 3915 + 3883 == 7798 ? True \n",
      "quick check: 3882 + 3875 == 7757 ? True \n",
      "quick check: 3895 + 3882 == 7777 ? True \n",
      "quick check: 3877 + 3874 == 7751 ? True \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n",
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n"
     ]
    }
   ],
   "source": [
    "# For each control set, only output MERGED version of ORF and Non-ORF with controls removed which don't map cleeanly\n",
    "for control_set in sorted(bedo.Control_Set.unique()):\n",
    "    # print(control_set)\n",
    "    df_temp = bedo[bedo.Control_Set == control_set].sort_values(by=['Chr','Gene_Start_hg38','Gene_End_hg38'])\n",
    "    df_temp_orf = df_temp[~df_temp.Status.str.contains('norf')]\n",
    "    df_temp_norf = df_temp[df_temp.Status.str.contains('norf')]\n",
    "    print(f'quick check: {df_temp_orf.shape[0]} + {df_temp_norf.shape[0]} == {df_temp.shape[0]} ? {df_temp_orf.shape[0] + df_temp_norf.shape[0] == df_temp.shape[0]} ')\n",
    "    df_by_set[control_set] = {'orf':bedify(df_temp_orf),'norf':bedify(df_temp_norf)}\n",
    "    bedify(df_temp).to_csv(f'grch38_intergenic_controls_combined_ORFs_non_ORFs_{control_set}_withGRCh37CrossRemoval.bed',sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6c7556c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/29/nl14y27n0516tj91qh8q3njh0000gq/T/ipykernel_57230/1873925465.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ret_df_2['zeroes'] = 0\n"
     ]
    }
   ],
   "source": [
    "# Also export one version with all ORFs and Non-ORFs from all control sets included \n",
    "\n",
    "df_temp = bedo.sort_values(by=['Chr','Gene_Start_hg38','Gene_End_hg38'])\n",
    "bedify(df_temp).to_csv(f'grch38_intergenic_controls_combined_ORFs_non_ORFs_all_control_sets_withGRCh37CrossRemoval.bed',sep='\\t',index=False,header=False,encoding=\"utf-8\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ae36ff40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>J</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-orf_6_control_set_4</th>\n",
       "      <td>TCAGCAGCATCGGGGGTCAGGAAAGACTTCACGAAGCCATAAATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:57850-58014:-norf_0_control_set_1</th>\n",
       "      <td>AAAACTAATCCCACATATAAACCCCTATGATAATTTCAGTTTGTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:59864-60013:-orf_8_control_set_5</th>\n",
       "      <td>TTAATTATCACTCACACTCCAAGACAAACACCATTTCAGTAGCAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632-67997.norf_segment:67741-67905:+norf_3_control_set_2</th>\n",
       "      <td>CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:108680-109563.norf_segment:108819-108959:+norf_9_control_set_5</th>\n",
       "      <td>TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26623039-26623182:-orf_0_control_set_1</th>\n",
       "      <td>TCACGCTGGCACCTTCTTGTGGGGATGGCATTGAGGACTAACACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26627091-26627363:-orf_1_control_set_1</th>\n",
       "      <td>TTATGGTAAATGGTGCAAGCAGTGGCAATGGGAATGGACTGCACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:56881824-56882003:-orf_0_control_set_1</th>\n",
       "      <td>TTAACACTGCTTCCTTCTCCCATCGAGAAGTAAAGCCCAGGTTCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070946-57071187.norf_segment:57070959-57071090:-norf_2_control_set_2</th>\n",
       "      <td>TATTCCATGCAGGTATCTTTCAATTGTAAATCATTACTGCAAAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:57209049-57209204:-orf_7_control_set_4</th>\n",
       "      <td>TCAGACACAGGTACAGCACATAGGCCAGGAGCCAGGGGGTGACTGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                    J\n",
       "Name                                                                                                 \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  TCAGCAGCATCGGGGGTCAGGAAAGACTTCACGAAGCCATAAATGC...\n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  AAAACTAATCCCACATATAAACCCCTATGATAATTTCAGTTTGTCC...\n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  TTAATTATCACTCACACTCCAAGACAAACACCATTTCAGTAGCAAT...\n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...\n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...\n",
       "...                                                                                               ...\n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  TCACGCTGGCACCTTCTTGTGGGGATGGCATTGAGGACTAACACTC...\n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  TTATGGTAAATGGTGCAAGCAGTGGCAATGGGAATGGACTGCACCC...\n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  TTAACACTGCTTCCTTCTCCCATCGAGAAGTAAAGCCCAGGTTCTG...\n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  TATTCCATGCAGGTATCTTTCAATTGTAAATCATTACTGCAAAAGA...\n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  TCAGACACAGGTACAGCACATAGGCCAGGAGCCAGGGGGTGACTGG...\n",
       "\n",
       "[38888 rows x 1 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# BEGIN second part of this notebook , generating genetic sequences for all of the controls \n",
    "# or genetic sequences (or reverse complement of them when appropriate, then amino acid sequences\n",
    "# Naming convention as this is a very sophomoric attempt at this\n",
    "# FASTAS derived from bedtools getfasta: https://bedtools.readthedocs.io/en/latest/content/tools/getfasta.html\n",
    "# and a downloaded version of GRCh38\n",
    "\n",
    "fastas = pd.read_csv('grch38_control_fastas.txt',sep='\\t',header=None)\n",
    "fastas_2 = fastas.rename(columns={0:'Name',1:'J'})#.set_index('Name')\n",
    "\n",
    "first_names = []\n",
    "for xi in fastas_2.Name:\n",
    "    xi = xi.split('::')[0]\n",
    "    first_names.append(xi)\n",
    "fastas_2['Name'] = first_names\n",
    "fastas_named = fastas_2.set_index('Name')\n",
    "fastas_named"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "db58c363",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>J</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-orf_6_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>129</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>TCAGCAGCATCGGGGGTCAGGAAAGACTTCACGAAGCCATAAATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:57850-58014:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>AAAACTAATCCCACATATAAACCCCTATGATAATTTCAGTTTGTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:59864-60013:-orf_8_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>150</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "      <td>TTAATTATCACTCACACTCCAAGACAAACACCATTTCAGTAGCAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632-67997.norf_segment:67741-67905:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:108680-109563.norf_segment:108819-108959:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>141</td>\n",
       "      <td>5</td>\n",
       "      <td>norf</td>\n",
       "      <td>TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26623039-26623182:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>TCACGCTGGCACCTTCTTGTGGGGATGGCATTGAGGACTAACACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26627091-26627363:-orf_1_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26627091</td>\n",
       "      <td>26627364</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>TTATGGTAAATGGTGCAAGCAGTGGCAATGGGAATGGACTGCACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:56881824-56882003:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>TTAACACTGCTTCCTTCTCCCATCGAGAAGTAAAGCCCAGGTTCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070946-57071187.norf_segment:57070959-57071090:-norf_2_control_set_2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57070959</td>\n",
       "      <td>57071091</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>132</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>TATTCCATGCAGGTATCTTTCAATTGTAAATCATTACTGCAAAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:57209049-57209204:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>156</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>TCAGACACAGGTACAGCACATAGGCCAGGAGCCAGGGGGTGACTGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1            35726   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  chr1            57850   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  chr1            59864   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  chr1            67741   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  chr1           108819   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26623039   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26627091   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  chrY         57070959   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  Quality  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          35855        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...          58015        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...          60014        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...          67906        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...         108960        0   \n",
       "...                                                           ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26623183        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26627364        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       56882004        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       57071091        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       57209205        0   \n",
       "\n",
       "                                                   Strand  Length Control_Set  \\\n",
       "Name                                                                            \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      -     129           4   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...      -     165           1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...      -     150           5   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...      +     165           2   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...      +     141           5   \n",
       "...                                                   ...     ...         ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     144           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     273           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     180           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...      -     132           2   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     156           4   \n",
       "\n",
       "                                                   Status  \\\n",
       "Name                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...   norf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...   norf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...   norf   \n",
       "...                                                   ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...   norf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...    orf   \n",
       "\n",
       "                                                                                                    J  \n",
       "Name                                                                                                   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  TCAGCAGCATCGGGGGTCAGGAAAGACTTCACGAAGCCATAAATGC...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  AAAACTAATCCCACATATAAACCCCTATGATAATTTCAGTTTGTCC...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  TTAATTATCACTCACACTCCAAGACAAACACCATTTCAGTAGCAAT...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...  \n",
       "...                                                                                               ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  TCACGCTGGCACCTTCTTGTGGGGATGGCATTGAGGACTAACACTC...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  TTATGGTAAATGGTGCAAGCAGTGGCAATGGGAATGGACTGCACCC...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  TTAACACTGCTTCCTTCTCCCATCGAGAAGTAAAGCCCAGGTTCTG...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  TATTCCATGCAGGTATCTTTCAATTGTAAATCATTACTGCAAAAGA...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  TCAGACACAGGTACAGCACATAGGCCAGGAGCCAGGGGGTGACTGG...  \n",
       "\n",
       "[38888 rows x 9 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df38 = bedo.join(fastas_named)\n",
    "df38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "82b7f65c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct for reverse complements\n",
    "for xi,yi in df38.iterrows():\n",
    "    if yi['Strand'] == '-':\n",
    "        df38.loc[xi,'J'] = Seq(yi['J']).reverse_complement().__str__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bf44e5b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>J</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:188174-188317:+orf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>188174</td>\n",
       "      <td>188318</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>144</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGTTGCTGGGAAGACCCCCAAGTCCCTCTTCTGCATCGTCCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:274832-275053:+orf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>274832</td>\n",
       "      <td>275054</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>222</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGACAGAAGTAATTCCTGAGTTGCTTCTGAAACCAGAGCTTCCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:363585-363872:+orf_5_control_set_3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>363585</td>\n",
       "      <td>363873</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>288</td>\n",
       "      <td>3</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGCTCCAAGCACATCCCACGGGGAGGACCATGAACAACTCAGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:384407-384607:+orf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>384407</td>\n",
       "      <td>384608</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>201</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGTGGTCTTATTCCCATAACTGGGGCCAGATGATCCCACCCTGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:789783-790127:+orf_2_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>789783</td>\n",
       "      <td>790128</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>345</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGGAATGGAAAGGACTCGAGTGGGATGGAATGGAGTGGAATGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26127501-26127659:+orf_3_control_set_2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26127501</td>\n",
       "      <td>26127660</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>159</td>\n",
       "      <td>2</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGAATGCTGATTGCCTGTACTCAACAGGTTTTCTGAAGAACTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26202219-26202383:+orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26202219</td>\n",
       "      <td>26202384</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGTTCAAATTATGTGGACCTTCCTGCCCACCCCCCGCAACCAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26588067-26588210:+orf_8_control_set_5</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26588067</td>\n",
       "      <td>26588211</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>144</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGCATTTGCATTCACGGCAGTGGGGTGTGATGGTGTTTGGTCCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26588359-26588487:+orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26588359</td>\n",
       "      <td>26588488</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>129</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGAGTTTCCCAATTGGGGAATCACCTATAAATAATCTTGAAATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26590842-26591003:+orf_6_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26590842</td>\n",
       "      <td>26591004</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>162</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGCTATGGTGTTATTTTTGCACTACAAGGGCTCTGAAGACTGCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9887 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...  chr1           188174   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...  chr1           274832   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1           363585   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1           384407   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:7...  chr1           789783   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26127501   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26202219   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26588067   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26588359   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26590842   \n",
       "\n",
       "                                                    Gene_End_hg38  Quality  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...         188318        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...         275054        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...         363873        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...         384608        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:7...         790128        0   \n",
       "...                                                           ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26127660        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26202384        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26588211        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26588488        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26591004        0   \n",
       "\n",
       "                                                   Strand  Length Control_Set  \\\n",
       "Name                                                                            \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...      +     144           1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...      +     222           1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      +     288           3   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      +     201           2   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:7...      +     345           2   \n",
       "...                                                   ...     ...         ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      +     159           2   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      +     165           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      +     144           5   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      +     129           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      +     162           4   \n",
       "\n",
       "                                                   Status  \\\n",
       "Name                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:7...    orf   \n",
       "...                                                   ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "\n",
       "                                                                                                    J  \n",
       "Name                                                                                                   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...  ATGTTGCTGGGAAGACCCCCAAGTCCCTCTTCTGCATCGTCCTCGG...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...  ATGACAGAAGTAATTCCTGAGTTGCTTCTGAAACCAGAGCTTCCCT...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  ATGCTCCAAGCACATCCCACGGGGAGGACCATGAACAACTCAGCTG...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  ATGTGGTCTTATTCCCATAACTGGGGCCAGATGATCCCACCCTGGG...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:7...  ATGGAATGGAAAGGACTCGAGTGGGATGGAATGGAGTGGAATGGAC...  \n",
       "...                                                                                               ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGAATGCTGATTGCCTGTACTCAACAGGTTTTCTGAAGAACTCAT...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGTTCAAATTATGTGGACCTTCCTGCCCACCCCCCGCAACCAAAA...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGCATTTGCATTCACGGCAGTGGGGTGTGATGGTGTTTGGTCCCT...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGAGTTTCCCAATTGGGGAATCACCTATAAATAATCTTGAAATAA...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGCTATGGTGTTATTTTTGCACTACAAGGGCTCTGAAGACTGCCA...  \n",
       "\n",
       "[9887 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "9887"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# brief check to assure everything starts with 'ATG' as DNA\n",
    "# Check that the count here is equal to the number of rows displayed\n",
    "df38plus = df38[(df38.Strand=='+')&(df38.Status.str[:3]=='orf')]\n",
    "display(df38plus)\n",
    "sum(df38plus[df38plus.J.str[:3].isin(['ATG','atg'])].Chr.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "467fc43f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>J</th>\n",
       "      <th>K</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-orf_6_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>129</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGGGTGCTGAAGCTCCCACGCCTGCCTGTGAAAATGGAGTCCTCT...</td>\n",
       "      <td>MGAEAPTPACENGVLSHLGEPGAAPRRMHLWLREVFPDPRCC*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:57850-58014:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "      <td>norf</td>\n",
       "      <td>AATAGGATGGAAATAGCTGAGATCAGACATCTCCTTTTCAGAGTGG...</td>\n",
       "      <td>NRMEIAEIRHLLFRVENEVYN*LESMKVRKLHIFKELHTRDKLKLS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:59864-60013:-orf_8_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>150</td>\n",
       "      <td>5</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGAAAGTGATATTTCATCTTATAAGAGCAATTCCAAAACAAAAGC...</td>\n",
       "      <td>MKVIFHLIRAIPKQKQLMSELLCHIWRFLLLLKFILLLKWCLSWSV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632-67997.norf_segment:67741-67905:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...</td>\n",
       "      <td>LADSNLQRCPNNTSLCLDSHHYPLPFSVFISAPKRDLYVKYCYTSA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:108680-109563.norf_segment:108819-108959:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>141</td>\n",
       "      <td>5</td>\n",
       "      <td>norf</td>\n",
       "      <td>TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...</td>\n",
       "      <td>FPVIYLVEMRPLIGMGRRSPFILMTQHG*TLSDYHCTLLYFQH*RCQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26623039-26623182:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGCTGTGCAGACCACGGCCTCCGCAGAGGATCCCCTCACCCAGGC...</td>\n",
       "      <td>MLCRPRPPQRIPSPRPQGLPSCSLRDEIFNNAVSVSPQCHPHKKVPA*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26627091-26627363:-orf_1_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26627091</td>\n",
       "      <td>26627364</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGCCGCGAGCACCCGCCTCCCCGCGCTCTGCTGGCCTCCACACCC...</td>\n",
       "      <td>MPRAPASPRSAGLHTLGPPFLSQGQRGVLEAPELHGSLGGRRSGVR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:56881824-56882003:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGGTAAAATTCATTGTGATCCTGACCTCTGATGCTGTCAGCCACA...</td>\n",
       "      <td>MVKFIVILTSDAVSHRKGQNNLRLKPCLSQSSGSPNMVSRWSLLLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070946-57071187.norf_segment:57070959-57071090:-norf_2_control_set_2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57070959</td>\n",
       "      <td>57071091</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>132</td>\n",
       "      <td>2</td>\n",
       "      <td>norf</td>\n",
       "      <td>GTTGGCACATCCTAAATTCACGACAGTCATGACTCAAGCACTCTTG...</td>\n",
       "      <td>VGTS*IHDSHDSSTLVLLLFLISYKDKICLLQ**FTIERYLHGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:57209049-57209204:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>156</td>\n",
       "      <td>4</td>\n",
       "      <td>orf</td>\n",
       "      <td>ATGGGGGTCAGGCAAGCTGACACCCGCTGTCCTGAGCCCATGTTCC...</td>\n",
       "      <td>MGVRQADTRCPEPMFLSHIIRGTACTVGSQASRAEPPQSPPGSWPM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1            35726   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  chr1            57850   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  chr1            59864   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  chr1            67741   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  chr1           108819   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26623039   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26627091   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  chrY         57070959   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  Quality  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          35855        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...          58015        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...          60014        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...          67906        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...         108960        0   \n",
       "...                                                           ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26623183        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26627364        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       56882004        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       57071091        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       57209205        0   \n",
       "\n",
       "                                                   Strand  Length Control_Set  \\\n",
       "Name                                                                            \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      -     129           4   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...      -     165           1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...      -     150           5   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...      +     165           2   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...      +     141           5   \n",
       "...                                                   ...     ...         ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     144           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     273           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     180           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...      -     132           2   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     156           4   \n",
       "\n",
       "                                                   Status  \\\n",
       "Name                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...   norf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...    orf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...   norf   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...   norf   \n",
       "...                                                   ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...    orf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...   norf   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...    orf   \n",
       "\n",
       "                                                                                                    J  \\\n",
       "Name                                                                                                    \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  ATGGGTGCTGAAGCTCCCACGCCTGCCTGTGAAAATGGAGTCCTCT...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  AATAGGATGGAAATAGCTGAGATCAGACATCTCCTTTTCAGAGTGG...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  ATGAAAGTGATATTTCATCTTATAAGAGCAATTCCAAAACAAAAGC...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...   \n",
       "...                                                                                               ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGCTGTGCAGACCACGGCCTCCGCAGAGGATCCCCTCACCCAGGC...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGCCGCGAGCACCCGCCTCCCCGCGCTCTGCTGGCCTCCACACCC...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  ATGGTAAAATTCATTGTGATCCTGACCTCTGATGCTGTCAGCCACA...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  GTTGGCACATCCTAAATTCACGACAGTCATGACTCAAGCACTCTTG...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  ATGGGGGTCAGGCAAGCTGACACCCGCTGTCCTGAGCCCATGTTCC...   \n",
       "\n",
       "                                                                                                    K  \n",
       "Name                                                                                                   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...        MGAEAPTPACENGVLSHLGEPGAAPRRMHLWLREVFPDPRCC*  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  NRMEIAEIRHLLFRVENEVYN*LESMKVRKLHIFKELHTRDKLKLS...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  MKVIFHLIRAIPKQKQLMSELLCHIWRFLLLLKFILLLKWCLSWSV...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  LADSNLQRCPNNTSLCLDSHHYPLPFSVFISAPKRDLYVKYCYTSA...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...    FPVIYLVEMRPLIGMGRRSPFILMTQHG*TLSDYHCTLLYFQH*RCQ  \n",
       "...                                                                                               ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...   MLCRPRPPQRIPSPRPQGLPSCSLRDEIFNNAVSVSPQCHPHKKVPA*  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  MPRAPASPRSAGLHTLGPPFLSQGQRGVLEAPELHGSLGGRRSGVR...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  MVKFIVILTSDAVSHRKGQNNLRLKPCLSQSSGSPNMVSRWSLLLR...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       VGTS*IHDSHDSSTLVLLLFLISYKDKICLLQ**FTIERYLHGI  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  MGVRQADTRCPEPMFLSHIIRGTACTVGSQASRAEPPQSPPGSWPM...  \n",
       "\n",
       "[38888 rows x 10 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Then return the amino acid sequence for these controls, 'K'\n",
    "df38['K'] = [Seq(dna).translate().__str__() for dna in df38['J']]\n",
    "df38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1f15929d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Status of which (ORF or NORF) and original control group it is \n",
    "for xi,yi in df38.iterrows():\n",
    "    df38.loc[xi,'Status'] = xi.split(':')[-1].split('_')[0][1:]+'_'+xi.split(':')[-1].split('_')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8f0614d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>Status</th>\n",
       "      <th>J</th>\n",
       "      <th>K</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-orf_6_control_set_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>129</td>\n",
       "      <td>4</td>\n",
       "      <td>orf_6</td>\n",
       "      <td>ATGGGTGCTGAAGCTCCCACGCCTGCCTGTGAAAATGGAGTCCTCT...</td>\n",
       "      <td>MGAEAPTPACENGVLSHLGEPGAAPRRMHLWLREVFPDPRCC*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608-59054.norf_segment:57850-58014:-norf_0_control_set_1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57850</td>\n",
       "      <td>58015</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "      <td>norf_0</td>\n",
       "      <td>AATAGGATGGAAATAGCTGAGATCAGACATCTCCTTTTCAGAGTGG...</td>\n",
       "      <td>NRMEIAEIRHLLFRVENEVYN*LESMKVRKLHIFKELHTRDKLKLS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:59864-60013:-orf_8_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>59864</td>\n",
       "      <td>60014</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>150</td>\n",
       "      <td>5</td>\n",
       "      <td>orf_8</td>\n",
       "      <td>ATGAAAGTGATATTTCATCTTATAAGAGCAATTCCAAAACAAAAGC...</td>\n",
       "      <td>MKVIFHLIRAIPKQKQLMSELLCHIWRFLLLLKFILLLKWCLSWSV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632-67997.norf_segment:67741-67905:+norf_3_control_set_2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67741</td>\n",
       "      <td>67906</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "      <td>norf_3</td>\n",
       "      <td>CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...</td>\n",
       "      <td>LADSNLQRCPNNTSLCLDSHHYPLPFSVFISAPKRDLYVKYCYTSA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:108680-109563.norf_segment:108819-108959:+norf_9_control_set_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108819</td>\n",
       "      <td>108960</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>141</td>\n",
       "      <td>5</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...</td>\n",
       "      <td>FPVIYLVEMRPLIGMGRRSPFILMTQHG*TLSDYHCTLLYFQH*RCQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26623039-26623182:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>ATGCTGTGCAGACCACGGCCTCCGCAGAGGATCCCCTCACCCAGGC...</td>\n",
       "      <td>MLCRPRPPQRIPSPRPQGLPSCSLRDEIFNNAVSVSPQCHPHKKVPA*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:26627091-26627363:-orf_1_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26627091</td>\n",
       "      <td>26627364</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273</td>\n",
       "      <td>1</td>\n",
       "      <td>orf_1</td>\n",
       "      <td>ATGCCGCGAGCACCCGCCTCCCCGCGCTCTGCTGGCCTCCACACCC...</td>\n",
       "      <td>MPRAPASPRSAGLHTLGPPFLSQGQRGVLEAPELHGSLGGRRSGVR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:56881824-56882003:-orf_0_control_set_1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>ATGGTAAAATTCATTGTGATCCTGACCTCTGATGCTGTCAGCCACA...</td>\n",
       "      <td>MVKFIVILTSDAVSHRKGQNNLRLKPCLSQSSGSPNMVSRWSLLLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070946-57071187.norf_segment:57070959-57071090:-norf_2_control_set_2</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57070959</td>\n",
       "      <td>57071091</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>132</td>\n",
       "      <td>2</td>\n",
       "      <td>norf_2</td>\n",
       "      <td>GTTGGCACATCCTAAATTCACGACAGTCATGACTCAAGCACTCTTG...</td>\n",
       "      <td>VGTS*IHDSHDSSTLVLLLFLISYKDKICLLQ**FTIERYLHGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:57209049-57209204:-orf_7_control_set_4</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209049</td>\n",
       "      <td>57209205</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>156</td>\n",
       "      <td>4</td>\n",
       "      <td>orf_7</td>\n",
       "      <td>ATGGGGGTCAGGCAAGCTGACACCCGCTGTCCTGAGCCCATGTTCC...</td>\n",
       "      <td>MGVRQADTRCPEPMFLSHIIRGTACTVGSQASRAEPPQSPPGSWPM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38888 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1            35726   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  chr1            57850   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  chr1            59864   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  chr1            67741   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  chr1           108819   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26623039   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  chrY         26627091   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         56881824   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  chrY         57070959   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  chrY         57209049   \n",
       "\n",
       "                                                    Gene_End_hg38  Quality  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          35855        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...          58015        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...          60014        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...          67906        0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...         108960        0   \n",
       "...                                                           ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26623183        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...       26627364        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       56882004        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       57071091        0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...       57209205        0   \n",
       "\n",
       "                                                   Strand  Length Control_Set  \\\n",
       "Name                                                                            \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...      -     129           4   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...      -     165           1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...      -     150           5   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...      +     165           2   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...      +     141           5   \n",
       "...                                                   ...     ...         ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     144           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...      -     273           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     180           1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...      -     132           2   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...      -     156           4   \n",
       "\n",
       "                                                    Status  \\\n",
       "Name                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...   orf_6   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  norf_0   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...   orf_8   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  norf_3   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  norf_9   \n",
       "...                                                    ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...   orf_0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...   orf_1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...   orf_0   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  norf_2   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...   orf_7   \n",
       "\n",
       "                                                                                                    J  \\\n",
       "Name                                                                                                    \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  ATGGGTGCTGAAGCTCCCACGCCTGCCTGTGAAAATGGAGTCCTCT...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  AATAGGATGGAAATAGCTGAGATCAGACATCTCCTTTTCAGAGTGG...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  ATGAAAGTGATATTTCATCTTATAAGAGCAATTCCAAAACAAAAGC...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  CTGGCAGATAGCAACTTACAAAGATGCCCCAACAATACCTCCTTGT...   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...  TTTCCAGTTATATATCTGGTAGAGATGAGGCCATTGATAGGAATGG...   \n",
       "...                                                                                               ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGCTGTGCAGACCACGGCCTCCGCAGAGGATCCCCTCACCCAGGC...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  ATGCCGCGAGCACCCGCCTCCCCGCGCTCTGCTGGCCTCCACACCC...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  ATGGTAAAATTCATTGTGATCCTGACCTCTGATGCTGTCAGCCACA...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...  GTTGGCACATCCTAAATTCACGACAGTCATGACTCAAGCACTCTTG...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  ATGGGGGTCAGGCAAGCTGACACCCGCTGTCCTGAGCCCATGTTCC...   \n",
       "\n",
       "                                                                                                    K  \n",
       "Name                                                                                                   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...        MGAEAPTPACENGVLSHLGEPGAAPRRMHLWLREVFPDPRCC*  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55608...  NRMEIAEIRHLLFRVENEVYN*LESMKVRKLHIFKELHTRDKLKLS...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:5...  MKVIFHLIRAIPKQKQLMSELLCHIWRFLLLLKFILLLKWCLSWSV...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:66632...  LADSNLQRCPNNTSLCLDSHHYPLPFSVFISAPKRDLYVKYCYTSA...  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:10868...    FPVIYLVEMRPLIGMGRRSPFILMTQHG*TLSDYHCTLLYFQH*RCQ  \n",
       "...                                                                                               ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...   MLCRPRPPQRIPSPRPQGLPSCSLRDEIFNNAVSVSPQCHPHKKVPA*  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:2...  MPRAPASPRSAGLHTLGPPFLSQGQRGVLEAPELHGSLGGRRSGVR...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  MVKFIVILTSDAVSHRKGQNNLRLKPCLSQSSGSPNMVSRWSLLLR...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:57070...       VGTS*IHDSHDSSTLVLLLFLISYKDKICLLQ**FTIERYLHGI  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:5...  MGVRQADTRCPEPMFLSHIIRGTACTVGSQASRAEPPQSPPGSWPM...  \n",
       "\n",
       "[38888 rows x 10 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "32577a55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# below cells are relevant export steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6e4329ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df38_orf = df38[(df38.Control_Set=='1')&(~df38.Status.str.contains('norf'))]\n",
    "df38_orf.to_csv(r'grch38_intergenic_controls_ORFs_set_1_with_sequences_withGRCh37CrossRemoval.tsv',sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6155577c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df38_norf = df38[(df38.Control_Set=='1')&(df38.Status.str.contains('norf'))]\n",
    "df38_norf.to_csv(r'grch38_intergenic_controls_non_ORFs_set_1_with_sequences_withGRCh37CrossRemoval.tsv',sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2d175545",
   "metadata": {},
   "outputs": [],
   "source": [
    "df38.to_csv(r'grch38_intergenic_controls_combined_ORFs_non_ORFs_all_sets_38888_with_sequences_withGRCh37CrossRemoval.tsv',sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "64b1066c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df38[df38.Control_Set=='1'].to_csv(r'grch38_intergenic_controls_combined_ORFs_non_ORFs_set_1_7805_with_sequences_withGRCh37CrossRemoval.tsv',sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83426709",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
