{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0fd5ecf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "## Daniel Marten\n",
    "## Creating GRCh38 BED files for controls\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "097c2f1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nam_controls: downloaded 'bam' files from Nam\n",
    "# 10 sets of GRCh38 controls, each with a pair of Intergenic ORF and Non-Intergenic ORFs\n",
    "# Make new dataframes and compress to FIVE control sets moving forward\n",
    "\n",
    "norf_dfs = []\n",
    "orf_dfs = []\n",
    "\n",
    "for gtfidx in range(10):\n",
    "    # Reading in all GRCh38 ORFs\n",
    "    orf_path = f'/Users/marten/Downloads/nam_controls_marten_June_2023/GRCh38_Ens89/sample-ORFs/GRCh38.Ens89.dna_rm.chromosome.all.intergenic_gt122.orfs.sample{gtfidx}.bed'\n",
    "    norf_path = f'/Users/marten/Downloads/nam_controls_marten_June_2023/GRCh38_Ens89/sample-nORFs/GRCh38.Ens89.dna_rm.chromosome.all.intergenic_gt122.norf_sample{gtfidx}.bed'\n",
    "    new_orf = pd.read_csv(orf_path,sep='\\t').reset_index()\n",
    "    new_norf = pd.read_csv(norf_path,sep='\\t').reset_index()\n",
    "    \n",
    "    # Formatting\n",
    "    col_renames = {'level_0':'Chr',\n",
    "               'level_1':'Gene_Start_hg38',\n",
    "               'level_2':'Gene_End_hg38',\n",
    "               'level_3':'Name',\n",
    "               '# All coordinates are 0-indexed, coordinates in ID column are end-inclusive, and coordinates in start/end columns are end-exclusive (as is conventional)':'Strand'\n",
    "              }\n",
    "    \n",
    "    new_orf = new_orf.rename(columns=col_renames)\n",
    "    new_orf = new_orf.set_index('Name').drop('level_4',axis=1)\n",
    "    new_orf['Status'] = [f'orf_{gtfidx}']*new_orf.shape[0]\n",
    "    new_orf['Control_Set'] = (gtfidx//2)+1 # THIS REDUCES THE 10 CONTROL GROUPS TO 5\n",
    "    #new_orf['annotation'] = ['intergenics']*new_orf.shape[0]\n",
    "    \n",
    "    new_norf = new_norf.rename(columns=col_renames)\n",
    "    new_norf = new_norf.set_index('Name').drop('level_4',axis=1)\n",
    "    new_norf['Status'] = [f'norf_{gtfidx}']*new_norf.shape[0]\n",
    "    new_norf['Control_Set'] = (gtfidx//2)+1 # THIS REDUCES THE 10 CONTROL GROUPS TO 5\n",
    "    #new_norf['annotation'] = ['intergenics']*new_norf.shape[0]\n",
    "\n",
    "    orf_dfs.append(new_orf)\n",
    "    norf_dfs.append(new_norf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99ca3de6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>level_1</th>\n",
       "      <th>level_2</th>\n",
       "      <th>level_3</th>\n",
       "      <th>level_4</th>\n",
       "      <th># All coordinates are 0-indexed, coordinates in ID column are end-inclusive, and coordinates in start/end columns are end-exclusive (as is conventional)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>188174</td>\n",
       "      <td>188318</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>274832</td>\n",
       "      <td>275054</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4879673</td>\n",
       "      <td>4880000</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4935350</td>\n",
       "      <td>4935473</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9453457</td>\n",
       "      <td>9453667</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1995</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21505338</td>\n",
       "      <td>21505467</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1996</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22446066</td>\n",
       "      <td>22446234</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1997</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24193329</td>\n",
       "      <td>24193533</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26623039</td>\n",
       "      <td>26623183</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881824</td>\n",
       "      <td>56882004</td>\n",
       "      <td>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     level_0   level_1   level_2  \\\n",
       "0       chr1    188174    188318   \n",
       "1       chr1    274832    275054   \n",
       "2       chr1   4879673   4880000   \n",
       "3       chr1   4935350   4935473   \n",
       "4       chr1   9453457   9453667   \n",
       "...      ...       ...       ...   \n",
       "1995    chrY  21505338  21505467   \n",
       "1996    chrY  22446066  22446234   \n",
       "1997    chrY  24193329  24193533   \n",
       "1998    chrY  26623039  26623183   \n",
       "1999    chrY  56881824  56882004   \n",
       "\n",
       "                                                level_3  level_4  \\\n",
       "0     GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...        0   \n",
       "1     GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...        0   \n",
       "2     GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...        0   \n",
       "3     GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...        0   \n",
       "4     GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:...        0   \n",
       "...                                                 ...      ...   \n",
       "1995  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...        0   \n",
       "1996  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...        0   \n",
       "1997  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...        0   \n",
       "1998  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...        0   \n",
       "1999  GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:...        0   \n",
       "\n",
       "     # All coordinates are 0-indexed, coordinates in ID column are end-inclusive, and coordinates in start/end columns are end-exclusive (as is conventional)  \n",
       "0                                                     +                                                                                                        \n",
       "1                                                     +                                                                                                        \n",
       "2                                                     +                                                                                                        \n",
       "3                                                     +                                                                                                        \n",
       "4                                                     +                                                                                                        \n",
       "...                                                 ...                                                                                                        \n",
       "1995                                                  -                                                                                                        \n",
       "1996                                                  -                                                                                                        \n",
       "1997                                                  -                                                                                                        \n",
       "1998                                                  -                                                                                                        \n",
       "1999                                                  -                                                                                                        \n",
       "\n",
       "[2000 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_bed = pd.read_csv(f'/Users/marten/Downloads/nam_controls_marten_June_2023/GRCh38_Ens89/sample-ORFs/GRCh38.Ens89.dna_rm.chromosome.all.intergenic_gt122.orfs.sample0.bed',\n",
    "                     sep='\\t').reset_index()\n",
    "raw_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3e007644",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:188174-188317:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>188174</td>\n",
       "      <td>188318</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:274832-275053:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>274832</td>\n",
       "      <td>275054</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4879673-4879999:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4879673</td>\n",
       "      <td>4880000</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4935350-4935472:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4935350</td>\n",
       "      <td>4935473</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9453457-9453666:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9453457</td>\n",
       "      <td>9453667</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902094-48904717.norf_segment:48903937-48904062:-</th>\n",
       "      <td>chr8</td>\n",
       "      <td>48903937</td>\n",
       "      <td>48904063</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:2747860-2748261.norf_segment:2748014-2748175:-</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2748014</td>\n",
       "      <td>2748176</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862538-55863365.norf_segment:55862998-55863138:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55862998</td>\n",
       "      <td>55863139</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:117634080-117635194.norf_segment:117634743-117635027:-</th>\n",
       "      <td>chr4</td>\n",
       "      <td>117634743</td>\n",
       "      <td>117635028</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654183-22655281.norf_segment:22654431-22654565:-</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22654431</td>\n",
       "      <td>22654566</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40000 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...  chr1           188174   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...  chr1           274832   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...  chr1          4879673   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...  chr1          4935350   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...  chr1          9453457   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...  chr8         48903937   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...  chr2          2748014   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...  chr1         55862998   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...  chr4        117634743   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...  chrY         22654431   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...         188318      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...         275054      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...        4880000      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...        4935473      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...        9453667      +   \n",
       "...                                                           ...    ...   \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...       48904063      -   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...        2748176      -   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...       55863139      -   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...      117635028      -   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...       22654566      -   \n",
       "\n",
       "                                                    Status  Control_Set  \n",
       "Name                                                                     \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...   orf_0            1  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...   orf_0            1  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...   orf_0            1  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...   orf_0            1  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...   orf_0            1  \n",
       "...                                                    ...          ...  \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...  norf_9            5  \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...  norf_9            5  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...  norf_9            5  \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...  norf_9            5  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...  norf_9            5  \n",
       "\n",
       "[40000 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Properly formatted dataframe of ALL control sequences - 40,000 of them \n",
    "control_list = orf_dfs + norf_dfs\n",
    "control_df = pd.concat(control_list)\n",
    "control_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebeda0ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:188174-188317:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>188174</td>\n",
       "      <td>188318</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "      <td>144</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:274832-275053:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>274832</td>\n",
       "      <td>275054</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "      <td>222</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4879673-4879999:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4879673</td>\n",
       "      <td>4880000</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "      <td>327</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4935350-4935472:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4935350</td>\n",
       "      <td>4935473</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "      <td>123</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9453457-9453666:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9453457</td>\n",
       "      <td>9453667</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "      <td>210</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902094-48904717.norf_segment:48903937-48904062:-</th>\n",
       "      <td>chr8</td>\n",
       "      <td>48903937</td>\n",
       "      <td>48904063</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>126</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:2747860-2748261.norf_segment:2748014-2748175:-</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2748014</td>\n",
       "      <td>2748176</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>162</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862538-55863365.norf_segment:55862998-55863138:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55862998</td>\n",
       "      <td>55863139</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>141</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:117634080-117635194.norf_segment:117634743-117635027:-</th>\n",
       "      <td>chr4</td>\n",
       "      <td>117634743</td>\n",
       "      <td>117635028</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>285</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654183-22655281.norf_segment:22654431-22654565:-</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22654431</td>\n",
       "      <td>22654566</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>135</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40000 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...  chr1           188174   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...  chr1           274832   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...  chr1          4879673   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...  chr1          4935350   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...  chr1          9453457   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...  chr8         48903937   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...  chr2          2748014   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...  chr1         55862998   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...  chr4        117634743   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...  chrY         22654431   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...         188318      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...         275054      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...        4880000      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...        4935473      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...        9453667      +   \n",
       "...                                                           ...    ...   \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...       48904063      -   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...        2748176      -   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...       55863139      -   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...      117635028      -   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...       22654566      -   \n",
       "\n",
       "                                                    Status  Control_Set  \\\n",
       "Name                                                                      \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...   orf_0            1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...   orf_0            1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...   orf_0            1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...   orf_0            1   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...   orf_0            1   \n",
       "...                                                    ...          ...   \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...  norf_9            5   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...  norf_9            5   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...  norf_9            5   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...  norf_9            5   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...  norf_9            5   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...          144    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...          222    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...          327    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...          123    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...          210    False   \n",
       "...                                                         ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...          126    False   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...          162    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...          141    False   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...          285    False   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...          135    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:1...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:2...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:4...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...          3  \n",
       "...                                                       ...  \n",
       "GRCh38.Ens89.dna_rm.chr8.intergenic_gt122:48902...          4  \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122:27478...          4  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122:55862...          4  \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122:11763...          4  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:22654...          4  \n",
       "\n",
       "[40000 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# User providing some annotations and length\n",
    "### HEIRARCHY\n",
    "# there are duplicates of many Intergenic ORFs between the different control sets, which we do not want\n",
    "# remove: duplicate ORFs between control sets \n",
    "# remove: any intergenic control sequences that overlap annotated or unannotated genes \n",
    "# heirarchy: annotated>unannotated>Igen-ORF>Igen-Non-ORF\n",
    "###\n",
    "# annotations for sorting later\n",
    "control_df['user_length'] = control_df.Gene_End_hg38 - control_df.Gene_Start_hg38\n",
    "control_df['removal'] = False\n",
    "control_df['heirarchy'] = 3\n",
    "control_df.loc[control_df['Status'].str.contains('norf'),'heirarchy'] = 4\n",
    "control_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "756d9eec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chrX\n",
      "chrY\n",
      "Total duplicates removed (instances):  108 (note that first occurence is kept)\n",
      "Total non-duplicate overlaps observed:  532\n",
      "Overlaps decided by heirarchy:  260\n",
      "Overlaps decided by length:  267\n",
      "Overlaps where the first occuring one is taken:  5\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(39370, 9)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# HERE: remove duplicate instances of intergenic ORFs\n",
    "# and instances of control sequences overlapping \n",
    "# decide by: heirarchy, then length, then whichever occurs first \n",
    "# keep Intergenic, longest, first-starting\n",
    "\n",
    "fixed_list = []\n",
    "dupl_list = []\n",
    "removed_list = []\n",
    "gi = 'Gene_Start_hg38'\n",
    "ge = 'Gene_End_hg38'\n",
    "\n",
    "total_dups = 0\n",
    "total_overlaps = 0\n",
    "by_heirarchy = 0\n",
    "by_length = 0\n",
    "punt = 0\n",
    "\n",
    "for unique_chrom in sorted(control_df.Chr.unique()):\n",
    "    print(unique_chrom)\n",
    "    by_chrom_df = control_df[control_df.Chr==unique_chrom]\n",
    "    \n",
    "    by_chrom_df = by_chrom_df.sort_values(by=['Chr',gi,ge])\n",
    "    \n",
    "    dupls = by_chrom_df[by_chrom_df.duplicated(['Chr','Gene_Start_hg38','Gene_End_hg38'],keep='first')]\n",
    "    dupl_list.append(dupls)\n",
    "    \n",
    "    by_chrom_df_dd = by_chrom_df[~by_chrom_df.duplicated(['Chr','Gene_Start_hg38','Gene_End_hg38'],keep='first')]\n",
    "    \n",
    "    total_dups += dupls.shape[0]\n",
    "    \n",
    "    for xi,yi in by_chrom_df_dd.iterrows():\n",
    "        \n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "        \n",
    "        query_new = by_chrom_df_dd[(by_chrom_df_dd[gi]>(index_start-1))&(by_chrom_df_dd[ge]<(index_end+1000000))]\n",
    "        \n",
    "        for xii,yii in query_new.iterrows():\n",
    "            \n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "            \n",
    "            case_1 = query_start in range(index_start,index_end+1)\n",
    "            case_2 = query_end in range(index_start,index_end+1)\n",
    "            case_3 = index_start in range(int(query_start),int(query_end)+1)\n",
    "            \n",
    "            if (xi!=xii):\n",
    "                if any([case_1,case_2,case_3]):\n",
    "                    # instance of overlap!\n",
    "                        total_overlaps += 1\n",
    "                        if yii['heirarchy'] > yi['heirarchy']:\n",
    "                            by_chrom_df_dd.loc[xii,'removal'] = True\n",
    "                            by_heirarchy += 1\n",
    "                        elif yii['heirarchy'] < yi['heirarchy']:\n",
    "                            by_chrom_df_dd.loc[xi,'removal'] = True\n",
    "                            by_heirarchy += 1\n",
    "                        else:\n",
    "                            if yii['user_length']>yi['user_length']:\n",
    "                                by_chrom_df_dd.loc[xi,'removal'] = True\n",
    "                                by_length += 1\n",
    "                            elif yii['user_length']<yi['user_length']:\n",
    "                                by_chrom_df_dd.loc[xii,'removal'] = True\n",
    "                                by_length += 1\n",
    "                            else:\n",
    "                                # Punt - slang to defer decision \n",
    "                                # decide which to keep based on which with earliest start\n",
    "                                punt += 1\n",
    "                                if yii[gi] > yi[gi]:\n",
    "                                    by_chrom_df_dd.loc[xii,'removal'] = True\n",
    "                                else:\n",
    "                                    by_chrom_df_dd.loc[xi,'removal'] = True\n",
    "     \n",
    "    removed_list.append(by_chrom_df_dd[by_chrom_df_dd.removal])\n",
    "    \n",
    "    by_chrom_df_dd = by_chrom_df_dd[~by_chrom_df_dd.removal]\n",
    "    fixed_list.append(by_chrom_df_dd)\n",
    "    \n",
    "print('Total duplicates removed (instances): ',total_dups, '(note that first occurence is kept)') # note that the first occurence is kept \n",
    "print('Total non-duplicate overlaps observed: ',total_overlaps)\n",
    "print('Overlaps decided by heirarchy: ',by_heirarchy)\n",
    "print('Overlaps decided by length: ',by_length)\n",
    "print('Overlaps where the first occuring one is taken: ',punt)\n",
    "\n",
    "fixed_df = pd.concat(fixed_list)\n",
    "\n",
    "fixed_df.shape\n",
    "            \n",
    "            \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "aca50077",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:381772-381948:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>381772</td>\n",
       "      <td>381949</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_5</td>\n",
       "      <td>3</td>\n",
       "      <td>177</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:389564-389698:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>389564</td>\n",
       "      <td>389699</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_5</td>\n",
       "      <td>3</td>\n",
       "      <td>135</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9286319-9286465:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9286319</td>\n",
       "      <td>9286466</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:30588657-30588800:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30588657</td>\n",
       "      <td>30588801</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_8</td>\n",
       "      <td>5</td>\n",
       "      <td>144</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:82185132-82185320:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>82185132</td>\n",
       "      <td>82185321</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_2</td>\n",
       "      <td>2</td>\n",
       "      <td>189</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:376685-376912:-</th>\n",
       "      <td>chrY</td>\n",
       "      <td>376685</td>\n",
       "      <td>376913</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:9803333-9804623.norf_segment:9803717-9803839:+</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9803717</td>\n",
       "      <td>9803840</td>\n",
       "      <td>+</td>\n",
       "      <td>norf_5</td>\n",
       "      <td>3</td>\n",
       "      <td>123</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:14131327-14132345.norf_segment:14131415-14131543:-</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14131415</td>\n",
       "      <td>14131544</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_3</td>\n",
       "      <td>2</td>\n",
       "      <td>129</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:14853739-14853882:+</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14853739</td>\n",
       "      <td>14853883</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_8</td>\n",
       "      <td>5</td>\n",
       "      <td>144</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:25987778-25988827.norf_segment:25988027-25988167:+</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25988027</td>\n",
       "      <td>25988168</td>\n",
       "      <td>+</td>\n",
       "      <td>norf_8</td>\n",
       "      <td>5</td>\n",
       "      <td>141</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>630 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1           381772   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1           389564   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...  chr1          9286319   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  chr1         30588657   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:8...  chr1         82185132   \n",
       "...                                                  ...              ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:3...  chrY           376685   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:98033...  chrY          9803717   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:14131...  chrY         14131415   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:1...  chrY         14853739   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:25987...  chrY         25988027   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...         381949      -   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...         389699      -   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...        9286466      -   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...       30588801      +   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:8...       82185321      -   \n",
       "...                                                           ...    ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:3...         376913      -   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:98033...        9803840      +   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:14131...       14131544      -   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:1...       14853883      +   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:25987...       25988168      +   \n",
       "\n",
       "                                                    Status  Control_Set  \\\n",
       "Name                                                                      \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...   orf_5            3   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...   orf_5            3   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...   orf_9            5   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...   orf_8            5   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:8...   orf_2            2   \n",
       "...                                                    ...          ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:3...   orf_0            1   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:98033...  norf_5            3   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:14131...  norf_3            2   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:1...   orf_8            5   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:25987...  norf_8            5   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          177    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          135    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...          147    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          144    False   \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:8...          189    False   \n",
       "...                                                         ...      ...   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:3...          228     True   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:98033...          123     True   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:14131...          129     True   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:1...          144     True   \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:25987...          141     True   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:9...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          3  \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:8...          3  \n",
       "...                                                       ...  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:3...          3  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:98033...          4  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:14131...          4  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122.orf:1...          3  \n",
       "GRCh38.Ens89.dna_rm.chrY.intergenic_gt122:25987...          4  \n",
       "\n",
       "[630 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print list of all removed controls \n",
    "internal_removal = pd.concat(dupl_list + removed_list)\n",
    "internal_removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fe736b6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:49124942-49125067:-</th>\n",
       "      <td>chr7</td>\n",
       "      <td>49124942</td>\n",
       "      <td>49125068</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_8</td>\n",
       "      <td>5</td>\n",
       "      <td>126</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:128215691-128216032:-</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128215691</td>\n",
       "      <td>128216033</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>342</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:4...  chr7         49124942   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  chr7        128215691   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:4...       49125068      -   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...      128216033      -   \n",
       "\n",
       "                                                   Status  Control_Set  \\\n",
       "Name                                                                     \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:4...  orf_8            5   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  orf_9            5   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:4...          126    False   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          342    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:4...          3  \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          3  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SANITY CHECK: looking at duplicates and making sure they're removed\n",
    "dupl_list[-5] # \"random\" removed duplicates by chromosome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2b0bddaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:128215691-128216032:-</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128215691</td>\n",
       "      <td>128216033</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_6</td>\n",
       "      <td>4</td>\n",
       "      <td>342</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:128215691-128216032:-</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128215691</td>\n",
       "      <td>128216033</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_9</td>\n",
       "      <td>5</td>\n",
       "      <td>342</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  chr7        128215691   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  chr7        128215691   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...      128216033      -   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...      128216033      -   \n",
       "\n",
       "                                                   Status  Control_Set  \\\n",
       "Name                                                                     \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  orf_6            4   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  orf_9            5   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          342    False   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          342    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          3  \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          3  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_df[(control_df.Chr=='chr7') & (control_df.Gene_Start_hg38==128215691)] # how the duplicate looks in original"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0611b852",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:128215691-128216032:-</th>\n",
       "      <td>chr7</td>\n",
       "      <td>128215691</td>\n",
       "      <td>128216033</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_6</td>\n",
       "      <td>4</td>\n",
       "      <td>342</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Chr  Gene_Start_hg38  \\\n",
       "Name                                                                        \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  chr7        128215691   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...      128216033      -   \n",
       "\n",
       "                                                   Status  Control_Set  \\\n",
       "Name                                                                     \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...  orf_6            4   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          342    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:1...          3  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fixed_df[(fixed_df.Chr=='chr7') & (fixed_df.Gene_Start_hg38==128215691)] # how the duplicate looks in final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ee090613",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:1846342-1846867.norf_segment:1846384-1846506:-</th>\n",
       "      <td>chr20</td>\n",
       "      <td>1846384</td>\n",
       "      <td>1846507</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_2</td>\n",
       "      <td>2</td>\n",
       "      <td>123</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:5280890-5281982.norf_segment:5281709-5281846:+</th>\n",
       "      <td>chr20</td>\n",
       "      <td>5281709</td>\n",
       "      <td>5281847</td>\n",
       "      <td>+</td>\n",
       "      <td>norf_1</td>\n",
       "      <td>1</td>\n",
       "      <td>138</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:7137549-7137689:-</th>\n",
       "      <td>chr20</td>\n",
       "      <td>7137549</td>\n",
       "      <td>7137690</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_8</td>\n",
       "      <td>5</td>\n",
       "      <td>141</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:22286316-22288224.norf_segment:22287159-22287422:+</th>\n",
       "      <td>chr20</td>\n",
       "      <td>22287159</td>\n",
       "      <td>22287423</td>\n",
       "      <td>+</td>\n",
       "      <td>norf_5</td>\n",
       "      <td>3</td>\n",
       "      <td>264</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:23162495-23163862.norf_segment:23163020-23163271:-</th>\n",
       "      <td>chr20</td>\n",
       "      <td>23163020</td>\n",
       "      <td>23163272</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_4</td>\n",
       "      <td>3</td>\n",
       "      <td>252</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:23162495-23163862.norf_segment:23163399-23163527:-</th>\n",
       "      <td>chr20</td>\n",
       "      <td>23163399</td>\n",
       "      <td>23163528</td>\n",
       "      <td>-</td>\n",
       "      <td>norf_7</td>\n",
       "      <td>4</td>\n",
       "      <td>129</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:63064440-63064715:+</th>\n",
       "      <td>chr20</td>\n",
       "      <td>63064440</td>\n",
       "      <td>63064716</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_1</td>\n",
       "      <td>1</td>\n",
       "      <td>276</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:1846...  chr20          1846384   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:5280...  chr20          5281709   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...  chr20          7137549   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...  chr20         22287159   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...  chr20         23163020   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...  chr20         23163399   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...  chr20         63064440   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:1846...        1846507      -   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:5280...        5281847      +   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...        7137690      -   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...       22287423      +   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...       23163272      -   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...       23163528      -   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...       63064716      +   \n",
       "\n",
       "                                                    Status  Control_Set  \\\n",
       "Name                                                                      \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:1846...  norf_2            2   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:5280...  norf_1            1   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...   orf_8            5   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...  norf_5            3   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...  norf_4            3   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...  norf_7            4   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...   orf_1            1   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:1846...          123     True   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:5280...          138     True   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...          141     True   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...          264     True   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...          252     True   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...          129     True   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...          276     True   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:1846...          4  \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:5280...          4  \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...          3  \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...          4  \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...          4  \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2316...          4  \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...          3  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SANITY CHECK - making sure removed things are actually removed \n",
    "removed_list[len(removed_list)//2] # chosen chromosome of removed samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "826d98e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:22286316-22288224.norf_segment:22287159-22287422:+</th>\n",
       "      <td>chr20</td>\n",
       "      <td>22287159</td>\n",
       "      <td>22287423</td>\n",
       "      <td>+</td>\n",
       "      <td>norf_5</td>\n",
       "      <td>3</td>\n",
       "      <td>264</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...  chr20         22287159   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...       22287423      +   \n",
       "\n",
       "                                                    Status  Control_Set  \\\n",
       "Name                                                                      \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...  norf_5            3   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...          264    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122:2228...          4  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_df[(control_df.Chr=='chr20') & (control_df.Gene_Start_hg38==22287159)] # sequences removed "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c26ac8f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chr, Gene_Start_hg38, Gene_End_hg38, Strand, Status, Control_Set, user_length, removal, heirarchy]\n",
       "Index: []"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fixed_df[(fixed_df.Chr=='chr20') & (fixed_df.Gene_Start_hg38==63064716)] \n",
    "# checking to make sure a random sequence is actually removed "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a1380488",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chrX\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "# Use the same code to check that, when ran again, no further sequences are removed \n",
    "\n",
    "gi = 'Gene_Start_hg38'\n",
    "ge = 'Gene_End_hg38'\n",
    "    \n",
    "for chrom_id in sorted(fixed_df.Chr.unique()):\n",
    "\n",
    "    print(chrom_id)\n",
    "\n",
    "    query_df = fixed_df[fixed_df['Chr']==chrom_id]\n",
    "\n",
    "    for xi,yi in query_df.iterrows():\n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "\n",
    "        query_new = query_df[(query_df[gi]>(index_start-1000000))&(query_df[ge]<(index_end+1000000))]\n",
    "\n",
    "        for xii,yii in query_new.iterrows():\n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "\n",
    "            if xi!=xii:\n",
    "                if query_start in range(index_start,index_end+1) or query_end in range(index_start,index_end+1):\n",
    "                    print(xi,yi)\n",
    "                    print(xii,yii) \n",
    "                    print('Exception(DANGER)')\n",
    "\n",
    "\n",
    "\n",
    "                elif index_start in range(int(query_start),int(query_end)+1):\n",
    "                    print(xi,yi)\n",
    "                    print(xii,yii) \n",
    "                    print('Exception(DANGER)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c8b93040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PS</th>\n",
       "      <th>Description</th>\n",
       "      <th>Plength</th>\n",
       "      <th>Gap_Gene?</th>\n",
       "      <th>Chr</th>\n",
       "      <th>OldLong_Transcript_Start_hg38</th>\n",
       "      <th>OldLong_Transcript_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>CDS_Start_hg38</th>\n",
       "      <th>CDS_End_hg38</th>\n",
       "      <th>Protein_Sequence</th>\n",
       "      <th>CDS_Sequence</th>\n",
       "      <th>in_old</th>\n",
       "      <th>Gene_Start</th>\n",
       "      <th>Gene_Stop</th>\n",
       "      <th>in_new</th>\n",
       "      <th>evoera_38</th>\n",
       "      <th>evoera5_38</th>\n",
       "      <th>annotation_38</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_1</th>\n",
       "      <td>31</td>\n",
       "      <td>ENST00000308604.5_18272_457</td>\n",
       "      <td>52</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>111429413</td>\n",
       "      <td>111429572</td>\n",
       "      <td>MTDTENHDSSPSSTSTCCPPITAGMQLKDSLGPGSNCPLWTLRPLH...</td>\n",
       "      <td>ATGACAGACACTGAAAATCACGACTCATCCCCCTCCAGCACCTCTA...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_10</th>\n",
       "      <td>31</td>\n",
       "      <td>ENST00000411630.2_23991_594</td>\n",
       "      <td>59</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>52713673</td>\n",
       "      <td>52713853</td>\n",
       "      <td>MLVATGQCSRCFMFTFSTFSFNCHNSEVDSVRDRLPQDHSAPANSM...</td>\n",
       "      <td>ATGCTGGTGGCAACAGGGCAGTGTAGCAGGTGCTTCATGTTCACCT...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_103</th>\n",
       "      <td>31</td>\n",
       "      <td>ENST00000499346.2_27384_333</td>\n",
       "      <td>84</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>128082767</td>\n",
       "      <td>128083022</td>\n",
       "      <td>MLGAFRSGPQPLPEPRARCVPQPGLLWALTRRRESPLVTPGLNLEE...</td>\n",
       "      <td>ATGCTGGGGGCTTTCCGGTCGGGGCCGCAGCCGCTTCCGGAGCCGC...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_108</th>\n",
       "      <td>24</td>\n",
       "      <td>ENST00000501177.3_12701_390</td>\n",
       "      <td>84</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>54919086</td>\n",
       "      <td>54925136</td>\n",
       "      <td>MLAEIHPKAGLQSLQFIMELLYWLLEGGDSEDKEDATGNVEMKNIQ...</td>\n",
       "      <td>ATGTTGGCTGAAATTCATCCCAAGGCTGGTCTGCAAAGTCTGCAAT...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baz_Hs_112</th>\n",
       "      <td>25</td>\n",
       "      <td>ENST00000503704.1_24073_293</td>\n",
       "      <td>52</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>82900166</td>\n",
       "      <td>82900435</td>\n",
       "      <td>MRSREAGPKLRRIQEPANGSPGAVSETGGYREERLSDAEIMGKLLA...</td>\n",
       "      <td>ATGCGAAGCAGAGAGGCAGGACCAAAATTGAGGCGAATCCAGGAAC...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_994</th>\n",
       "      <td>31</td>\n",
       "      <td>HF584391</td>\n",
       "      <td>69</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>143628143</td>\n",
       "      <td>143628353</td>\n",
       "      <td>MLYTHNTEFNLKRQICFVPQCKTFVSLCFVKQTQENWYTCTSWVLY...</td>\n",
       "      <td>ATGCTTTATACACATAATACTGAATTTAACCTCAAGAGGCAAATCT...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_995</th>\n",
       "      <td>31</td>\n",
       "      <td>HF583960</td>\n",
       "      <td>46</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>104020518</td>\n",
       "      <td>104021665</td>\n",
       "      <td>MREWLSIRNMRIKCEIFSCSVKPMSANCISCRMKNATCWLSMRLRN</td>\n",
       "      <td>ATGAGAGAATGGCTCAGCATCAGAAACATGAGAATCAAATGCGAGA...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_997</th>\n",
       "      <td>31</td>\n",
       "      <td>HF548108</td>\n",
       "      <td>40</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>70860637</td>\n",
       "      <td>70860760</td>\n",
       "      <td>MFAYKGSSYHVSNTSNSINPTPKLASNPVGRYCMIKCLII</td>\n",
       "      <td>ATGTTTGCATATAAGGGAAGTAGTTATCATGTTAGTAATACCTCTA...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_998</th>\n",
       "      <td>31</td>\n",
       "      <td>HF583700</td>\n",
       "      <td>82</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>50099104</td>\n",
       "      <td>50099353</td>\n",
       "      <td>MLLVQGQHQNEEGLTRHLLSSSFTLSLPTPSFPLPHKVPMCLYPPL...</td>\n",
       "      <td>ATGCTGTTGGTTCAAGGACAACACCAGAATGAAGAGGGTCTCACAA...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vdp2013_S4_999</th>\n",
       "      <td>31</td>\n",
       "      <td>HF548109</td>\n",
       "      <td>60</td>\n",
       "      <td>Not_Gap_Gene</td>\n",
       "      <td>chr18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2718365</td>\n",
       "      <td>2722632</td>\n",
       "      <td>MKVMQTLMQEDHYHLKQLSFLLKRVSQRNFHLVFWIFLFVLEFHLI...</td>\n",
       "      <td>ATGAAAGTAATGCAGACACTTATGCAGGAAGACCACTACCATCTAA...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>4-Primate</td>\n",
       "      <td>5-Primate</td>\n",
       "      <td>Unannotated With Overlap</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21436 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                PS                  Description  Plength     Gap_Gene?    Chr  \\\n",
       "Name                                                                            \n",
       "Baz_Hs_1        31  ENST00000308604.5_18272_457       52  Not_Gap_Gene   chr2   \n",
       "Baz_Hs_10       31  ENST00000411630.2_23991_594       59  Not_Gap_Gene   chr4   \n",
       "Baz_Hs_103      31  ENST00000499346.2_27384_333       84  Not_Gap_Gene   chr5   \n",
       "Baz_Hs_108      24  ENST00000501177.3_12701_390       84  Not_Gap_Gene  chr16   \n",
       "Baz_Hs_112      25  ENST00000503704.1_24073_293       52  Not_Gap_Gene   chr4   \n",
       "...             ..                          ...      ...           ...    ...   \n",
       "vdp2013_S4_994  31                     HF584391       69  Not_Gap_Gene   chrX   \n",
       "vdp2013_S4_995  31                     HF583960       46  Not_Gap_Gene  chr10   \n",
       "vdp2013_S4_997  31                     HF548108       40  Not_Gap_Gene   chr6   \n",
       "vdp2013_S4_998  31                     HF583700       82  Not_Gap_Gene  chr12   \n",
       "vdp2013_S4_999  31                     HF548109       60  Not_Gap_Gene  chr18   \n",
       "\n",
       "                OldLong_Transcript_Start_hg38  OldLong_Transcript_End_hg38  \\\n",
       "Name                                                                         \n",
       "Baz_Hs_1                                  NaN                          NaN   \n",
       "Baz_Hs_10                                 NaN                          NaN   \n",
       "Baz_Hs_103                                NaN                          NaN   \n",
       "Baz_Hs_108                                NaN                          NaN   \n",
       "Baz_Hs_112                                NaN                          NaN   \n",
       "...                                       ...                          ...   \n",
       "vdp2013_S4_994                            NaN                          NaN   \n",
       "vdp2013_S4_995                            NaN                          NaN   \n",
       "vdp2013_S4_997                            NaN                          NaN   \n",
       "vdp2013_S4_998                            NaN                          NaN   \n",
       "vdp2013_S4_999                            NaN                          NaN   \n",
       "\n",
       "               Strand  CDS_Start_hg38  CDS_End_hg38  \\\n",
       "Name                                                  \n",
       "Baz_Hs_1            -       111429413     111429572   \n",
       "Baz_Hs_10           +        52713673      52713853   \n",
       "Baz_Hs_103          -       128082767     128083022   \n",
       "Baz_Hs_108          -        54919086      54925136   \n",
       "Baz_Hs_112          -        82900166      82900435   \n",
       "...               ...             ...           ...   \n",
       "vdp2013_S4_994      -       143628143     143628353   \n",
       "vdp2013_S4_995      +       104020518     104021665   \n",
       "vdp2013_S4_997      +        70860637      70860760   \n",
       "vdp2013_S4_998      +        50099104      50099353   \n",
       "vdp2013_S4_999      +         2718365       2722632   \n",
       "\n",
       "                                                 Protein_Sequence  \\\n",
       "Name                                                                \n",
       "Baz_Hs_1        MTDTENHDSSPSSTSTCCPPITAGMQLKDSLGPGSNCPLWTLRPLH...   \n",
       "Baz_Hs_10       MLVATGQCSRCFMFTFSTFSFNCHNSEVDSVRDRLPQDHSAPANSM...   \n",
       "Baz_Hs_103      MLGAFRSGPQPLPEPRARCVPQPGLLWALTRRRESPLVTPGLNLEE...   \n",
       "Baz_Hs_108      MLAEIHPKAGLQSLQFIMELLYWLLEGGDSEDKEDATGNVEMKNIQ...   \n",
       "Baz_Hs_112      MRSREAGPKLRRIQEPANGSPGAVSETGGYREERLSDAEIMGKLLA...   \n",
       "...                                                           ...   \n",
       "vdp2013_S4_994  MLYTHNTEFNLKRQICFVPQCKTFVSLCFVKQTQENWYTCTSWVLY...   \n",
       "vdp2013_S4_995     MREWLSIRNMRIKCEIFSCSVKPMSANCISCRMKNATCWLSMRLRN   \n",
       "vdp2013_S4_997           MFAYKGSSYHVSNTSNSINPTPKLASNPVGRYCMIKCLII   \n",
       "vdp2013_S4_998  MLLVQGQHQNEEGLTRHLLSSSFTLSLPTPSFPLPHKVPMCLYPPL...   \n",
       "vdp2013_S4_999  MKVMQTLMQEDHYHLKQLSFLLKRVSQRNFHLVFWIFLFVLEFHLI...   \n",
       "\n",
       "                                                     CDS_Sequence  in_old  \\\n",
       "Name                                                                        \n",
       "Baz_Hs_1        ATGACAGACACTGAAAATCACGACTCATCCCCCTCCAGCACCTCTA...    True   \n",
       "Baz_Hs_10       ATGCTGGTGGCAACAGGGCAGTGTAGCAGGTGCTTCATGTTCACCT...    True   \n",
       "Baz_Hs_103      ATGCTGGGGGCTTTCCGGTCGGGGCCGCAGCCGCTTCCGGAGCCGC...    True   \n",
       "Baz_Hs_108      ATGTTGGCTGAAATTCATCCCAAGGCTGGTCTGCAAAGTCTGCAAT...    True   \n",
       "Baz_Hs_112      ATGCGAAGCAGAGAGGCAGGACCAAAATTGAGGCGAATCCAGGAAC...    True   \n",
       "...                                                           ...     ...   \n",
       "vdp2013_S4_994  ATGCTTTATACACATAATACTGAATTTAACCTCAAGAGGCAAATCT...    True   \n",
       "vdp2013_S4_995  ATGAGAGAATGGCTCAGCATCAGAAACATGAGAATCAAATGCGAGA...    True   \n",
       "vdp2013_S4_997  ATGTTTGCATATAAGGGAAGTAGTTATCATGTTAGTAATACCTCTA...    True   \n",
       "vdp2013_S4_998  ATGCTGTTGGTTCAAGGACAACACCAGAATGAAGAGGGTCTCACAA...    True   \n",
       "vdp2013_S4_999  ATGAAAGTAATGCAGACACTTATGCAGGAAGACCACTACCATCTAA...    True   \n",
       "\n",
       "                Gene_Start  Gene_Stop  in_new  evoera_38 evoera5_38  \\\n",
       "Name                                                                  \n",
       "Baz_Hs_1               NaN        NaN    True  4-Primate  5-Primate   \n",
       "Baz_Hs_10              NaN        NaN    True  4-Primate  5-Primate   \n",
       "Baz_Hs_103             NaN        NaN    True  4-Primate  5-Primate   \n",
       "Baz_Hs_108             NaN        NaN    True  4-Primate  5-Primate   \n",
       "Baz_Hs_112             NaN        NaN    True  4-Primate  5-Primate   \n",
       "...                    ...        ...     ...        ...        ...   \n",
       "vdp2013_S4_994         NaN        NaN    True  4-Primate  5-Primate   \n",
       "vdp2013_S4_995         NaN        NaN    True  4-Primate  5-Primate   \n",
       "vdp2013_S4_997         NaN        NaN    True  4-Primate  5-Primate   \n",
       "vdp2013_S4_998         NaN        NaN    True  4-Primate  5-Primate   \n",
       "vdp2013_S4_999         NaN        NaN    True  4-Primate  5-Primate   \n",
       "\n",
       "                           annotation_38  \n",
       "Name                                      \n",
       "Baz_Hs_1        Unannotated With Overlap  \n",
       "Baz_Hs_10       Unannotated With Overlap  \n",
       "Baz_Hs_103      Unannotated With Overlap  \n",
       "Baz_Hs_108      Unannotated With Overlap  \n",
       "Baz_Hs_112      Unannotated With Overlap  \n",
       "...                                  ...  \n",
       "vdp2013_S4_994  Unannotated With Overlap  \n",
       "vdp2013_S4_995  Unannotated With Overlap  \n",
       "vdp2013_S4_997  Unannotated With Overlap  \n",
       "vdp2013_S4_998  Unannotated With Overlap  \n",
       "vdp2013_S4_999  Unannotated With Overlap  \n",
       "\n",
       "[21436 rows x 19 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# internal duplicates taken care of, now time to do the external (annotated and unannotated genes)\n",
    "phylo_df = pd.read_csv('/Users/marten/ug-gc/marten_completeGRCh38_21436genes_47UGremoved_gene_transcript_cds_metadata_mashup_old_new_diffFix_20231005.tsv',sep='\\t',index_col='Name')\n",
    "phylo_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e65e71ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set one column with length information\n",
    "# Unannotated Genes do not have 'oldest longest transcript' information\n",
    "# so set their CDS_Start and CDS_End information there instead\n",
    "\n",
    "for xi,yi in phylo_df.iterrows():\n",
    "    if str(yi['OldLong_Transcript_Start_hg38'] == 'nan'):\n",
    "        phylo_df.loc[xi,'OldLong_Transcript_Start_hg38'] = phylo_df.loc[xi,'CDS_Start_hg38']\n",
    "        phylo_df.loc[xi,'OldLong_Transcript_End_hg38'] = phylo_df.loc[xi,'CDS_End_hg38']\n",
    "    if 'ENSP' in xi[:4]:\n",
    "        phylo_df.loc[xi,'Status'] = 'annotated'\n",
    "    else:\n",
    "        phylo_df.loc[xi,'Status'] = 'unannotated'\n",
    "    phylo_df.loc[xi,'Control_Set'] = 'victorgenes' # known annotated & unannotated genes\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bf712380",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename this to 'Gene Start' and 'Gene End'\n",
    "phylo_df = phylo_df.rename(columns={'OldLong_Transcript_Start_hg38':'Gene_Start_hg38','OldLong_Transcript_End_hg38':'Gene_End_hg38'})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5db34c2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chrX\n",
      "chrY\n",
      "Total ORF/NORFs removed due to overlap with known annotated&unannoted genes:  201\n",
      "For UG:  200\n",
      "For Ensembl:  1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(39177, 9)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Code to: remove control sequences which overlap known annotated and unannotated genes\n",
    "\n",
    "ug_cleaned_dfs = []\n",
    "ugag_removal_list = []\n",
    "\n",
    "known_overlap = 0\n",
    "ug_overlap = 0\n",
    "ag_overlap = 0\n",
    "\n",
    "for unique_chrom in sorted(fixed_df.Chr.unique()):\n",
    "    print(unique_chrom)\n",
    "    by_chrom_df = fixed_df[fixed_df.Chr==unique_chrom]\n",
    "    by_chrom_df_dd = by_chrom_df.sort_values(by=[gi,ge])\n",
    "    \n",
    "    phylo_by_chr = phylo_df[phylo_df.Chr==unique_chrom].sort_values(by=[gi,ge])\n",
    "    \n",
    "    total_dups += dupls.shape[0]\n",
    "    \n",
    "    for xi,yi in by_chrom_df_dd.iterrows():\n",
    "        \n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "        \n",
    "        query_new = phylo_by_chr[(phylo_by_chr[gi]>(index_start-1000000))&(phylo_by_chr[ge]<(index_end+1000000))]\n",
    "        \n",
    "        for xii,yii in query_new.iterrows():\n",
    "            \n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "            \n",
    "            case_1 = query_start in range(index_start,index_end+1)\n",
    "            case_2 = query_end in range(index_start,index_end+1)\n",
    "            case_3 = index_start in range(int(query_start),int(query_end)+1)\n",
    "            \n",
    "            if (xi!=xii):\n",
    "                if any([case_1,case_2,case_3]):\n",
    "                    by_chrom_df_dd.loc[xi,'removal']=True\n",
    "                    known_overlap += 1\n",
    "                    \n",
    "                    if yii['Status']=='annotated':\n",
    "                        ag_overlap += 1\n",
    "                    elif yii['Status']=='unannotated':\n",
    "                        ug_overlap += 1\n",
    "                    else:\n",
    "                        raise Exception('Exception')\n",
    "\n",
    "     \n",
    "    ugag_removal_list.append(by_chrom_df_dd[by_chrom_df_dd.removal])\n",
    "    by_chrom_df_dd = by_chrom_df_dd[~by_chrom_df_dd.removal]\n",
    "    ug_cleaned_dfs.append(by_chrom_df_dd)\n",
    "    \n",
    "print('Total ORF/NORFs removed due to overlap with known annotated&unannoted genes: ',known_overlap)\n",
    "print('For UG: ',ug_overlap)\n",
    "print('For Ensembl: ',ag_overlap)\n",
    "\n",
    "df2 = pd.concat(ug_cleaned_dfs)\n",
    "\n",
    "df2.shape   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "158ed2e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr2\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chrX\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "# Run the same code again to ensure that nothing is removed the second time \n",
    "\n",
    "for chrom_id in df2.Chr.unique():\n",
    "\n",
    "    print(chrom_id)\n",
    "\n",
    "    query_df = df2[df2['Chr']==chrom_id]\n",
    "\n",
    "    for xi,yi in query_df.iterrows():\n",
    "        index_start = yi[gi]\n",
    "        index_end = yi[ge]\n",
    "        \n",
    "        p_by_c = phylo_df[phylo_df.Chr==chrom_id]\n",
    "        query_new = p_by_c[(p_by_c[gi]>(index_start-1000000))&(p_by_c[ge]<(index_end+1000000))]\n",
    "\n",
    "        for xii,yii in query_new.iterrows():\n",
    "            query_start = yii[gi]\n",
    "            query_end = yii[ge]\n",
    "\n",
    "            if xi!=xii:\n",
    "                if query_start in range(index_start,index_end+1) or query_end in range(index_start,index_end+1):\n",
    "                    print(xi,yi)\n",
    "                    print(xii,yii) \n",
    "                    raise Exception('DANGER')\n",
    "\n",
    "\n",
    "\n",
    "                elif index_start in range(int(query_start),int(query_end)+1):\n",
    "                    raise Exception('DANGER')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "991d1b6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chr                   chr17\n",
       "Gene_Start_hg38    83055202\n",
       "Gene_End_hg38      83055523\n",
       "Strand                    -\n",
       "Status               norf_0\n",
       "Control_Set               1\n",
       "user_length             321\n",
       "removal                True\n",
       "heirarchy                 4\n",
       "Name: GRCh38.Ens89.dna_rm.chr17.intergenic_gt122:83054956-83057066.norf_segment:83055202-83055522:-, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a sequence that should've been removed \n",
    "ugag_removal_list[len(ugag_removal_list)//3].iloc[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d3dad507",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chr, Gene_Start_hg38, Gene_End_hg38, Strand, Status, Control_Set, user_length, removal, heirarchy]\n",
       "Index: []"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# making sure that it is actually removed \n",
    "df2[(df2.Chr=='chr17')&(df2.Gene_Start_hg38==83055202)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a5b8c23a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to turn the DFs as we have them formatted into properly formatted BEDs with the control set in their name\n",
    "\n",
    "def bedify(input_df):\n",
    "    ret_df = df2.reset_index()\n",
    "    newnames = [ret_df.loc[xi,'Name']+ret_df.loc[xi,'Status']+'_control_set_'+str(ret_df.loc[xi,'Control_Set']) for xi in ret_df.index] \n",
    "    ret_df['new_index'] = newnames\n",
    "    ret_df_2 = ret_df[['Chr','Gene_Start_hg38','Gene_End_hg38','new_index','Strand']]\n",
    "    ret_df_2['zeroes'] = 0\n",
    "    ret_df_2 = ret_df_2[['Chr','Gene_Start_hg38','Gene_End_hg38','new_index','zeroes','Strand']]\n",
    "    ret_df_2.new_index[1]\n",
    "    return ret_df_2\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ea91a6cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export entire set\n",
    "bedify(df2).to_csv(r'grch38_bed_complete_allsets.bed',sep='\\t',index=False,header=False)\n",
    "# Create one dataframe per set of GRCh38 Controls\n",
    "df_by_set = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c1bc00ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Gene_Start_hg38</th>\n",
       "      <th>Gene_End_hg38</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Status</th>\n",
       "      <th>Control_Set</th>\n",
       "      <th>user_length</th>\n",
       "      <th>removal</th>\n",
       "      <th>heirarchy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:35726-35854:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35726</td>\n",
       "      <td>35855</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_6</td>\n",
       "      <td>4</td>\n",
       "      <td>129</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr10.intergenic_gt122.orf:78457018-78457278:+</th>\n",
       "      <td>chr10</td>\n",
       "      <td>78457018</td>\n",
       "      <td>78457279</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_1</td>\n",
       "      <td>1</td>\n",
       "      <td>261</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr13.intergenic_gt122.orf:19295575-19295712:+</th>\n",
       "      <td>chr13</td>\n",
       "      <td>19295575</td>\n",
       "      <td>19295713</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_1</td>\n",
       "      <td>1</td>\n",
       "      <td>138</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr15.intergenic_gt122.orf:96791710-96791832:+</th>\n",
       "      <td>chr15</td>\n",
       "      <td>96791710</td>\n",
       "      <td>96791833</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_4</td>\n",
       "      <td>3</td>\n",
       "      <td>123</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr2.intergenic_gt122.orf:1612891-1613037:-</th>\n",
       "      <td>chr2</td>\n",
       "      <td>1612891</td>\n",
       "      <td>1613038</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_2</td>\n",
       "      <td>2</td>\n",
       "      <td>147</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:24084104-24084259:-</th>\n",
       "      <td>chr20</td>\n",
       "      <td>24084104</td>\n",
       "      <td>24084260</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_4</td>\n",
       "      <td>3</td>\n",
       "      <td>156</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr4.intergenic_gt122.orf:12441312-12441443:-</th>\n",
       "      <td>chr4</td>\n",
       "      <td>12441312</td>\n",
       "      <td>12441444</td>\n",
       "      <td>-</td>\n",
       "      <td>orf_3</td>\n",
       "      <td>2</td>\n",
       "      <td>132</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr5.intergenic_gt122.orf:121017800-121017922:+</th>\n",
       "      <td>chr5</td>\n",
       "      <td>121017800</td>\n",
       "      <td>121017923</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_6</td>\n",
       "      <td>4</td>\n",
       "      <td>123</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:55593255-55593437:+</th>\n",
       "      <td>chr7</td>\n",
       "      <td>55593255</td>\n",
       "      <td>55593438</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_0</td>\n",
       "      <td>1</td>\n",
       "      <td>183</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38.Ens89.dna_rm.chr9.intergenic_gt122.orf:62847254-62847424:+</th>\n",
       "      <td>chr9</td>\n",
       "      <td>62847254</td>\n",
       "      <td>62847425</td>\n",
       "      <td>+</td>\n",
       "      <td>orf_1</td>\n",
       "      <td>1</td>\n",
       "      <td>171</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Chr  Gene_Start_hg38  \\\n",
       "Name                                                                         \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...   chr1            35726   \n",
       "GRCh38.Ens89.dna_rm.chr10.intergenic_gt122.orf:...  chr10         78457018   \n",
       "GRCh38.Ens89.dna_rm.chr13.intergenic_gt122.orf:...  chr13         19295575   \n",
       "GRCh38.Ens89.dna_rm.chr15.intergenic_gt122.orf:...  chr15         96791710   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122.orf:1...   chr2          1612891   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...  chr20         24084104   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122.orf:1...   chr4         12441312   \n",
       "GRCh38.Ens89.dna_rm.chr5.intergenic_gt122.orf:1...   chr5        121017800   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:5...   chr7         55593255   \n",
       "GRCh38.Ens89.dna_rm.chr9.intergenic_gt122.orf:6...   chr9         62847254   \n",
       "\n",
       "                                                    Gene_End_hg38 Strand  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          35855      -   \n",
       "GRCh38.Ens89.dna_rm.chr10.intergenic_gt122.orf:...       78457279      +   \n",
       "GRCh38.Ens89.dna_rm.chr13.intergenic_gt122.orf:...       19295713      +   \n",
       "GRCh38.Ens89.dna_rm.chr15.intergenic_gt122.orf:...       96791833      +   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122.orf:1...        1613038      -   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...       24084260      -   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122.orf:1...       12441444      -   \n",
       "GRCh38.Ens89.dna_rm.chr5.intergenic_gt122.orf:1...      121017923      +   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:5...       55593438      +   \n",
       "GRCh38.Ens89.dna_rm.chr9.intergenic_gt122.orf:6...       62847425      +   \n",
       "\n",
       "                                                   Status  Control_Set  \\\n",
       "Name                                                                     \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...  orf_6            4   \n",
       "GRCh38.Ens89.dna_rm.chr10.intergenic_gt122.orf:...  orf_1            1   \n",
       "GRCh38.Ens89.dna_rm.chr13.intergenic_gt122.orf:...  orf_1            1   \n",
       "GRCh38.Ens89.dna_rm.chr15.intergenic_gt122.orf:...  orf_4            3   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122.orf:1...  orf_2            2   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...  orf_4            3   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122.orf:1...  orf_3            2   \n",
       "GRCh38.Ens89.dna_rm.chr5.intergenic_gt122.orf:1...  orf_6            4   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:5...  orf_0            1   \n",
       "GRCh38.Ens89.dna_rm.chr9.intergenic_gt122.orf:6...  orf_1            1   \n",
       "\n",
       "                                                    user_length  removal  \\\n",
       "Name                                                                       \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          129    False   \n",
       "GRCh38.Ens89.dna_rm.chr10.intergenic_gt122.orf:...          261    False   \n",
       "GRCh38.Ens89.dna_rm.chr13.intergenic_gt122.orf:...          138    False   \n",
       "GRCh38.Ens89.dna_rm.chr15.intergenic_gt122.orf:...          123    False   \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122.orf:1...          147    False   \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...          156    False   \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122.orf:1...          132    False   \n",
       "GRCh38.Ens89.dna_rm.chr5.intergenic_gt122.orf:1...          123    False   \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:5...          183    False   \n",
       "GRCh38.Ens89.dna_rm.chr9.intergenic_gt122.orf:6...          171    False   \n",
       "\n",
       "                                                    heirarchy  \n",
       "Name                                                           \n",
       "GRCh38.Ens89.dna_rm.chr1.intergenic_gt122.orf:3...          3  \n",
       "GRCh38.Ens89.dna_rm.chr10.intergenic_gt122.orf:...          3  \n",
       "GRCh38.Ens89.dna_rm.chr13.intergenic_gt122.orf:...          3  \n",
       "GRCh38.Ens89.dna_rm.chr15.intergenic_gt122.orf:...          3  \n",
       "GRCh38.Ens89.dna_rm.chr2.intergenic_gt122.orf:1...          3  \n",
       "GRCh38.Ens89.dna_rm.chr20.intergenic_gt122.orf:...          3  \n",
       "GRCh38.Ens89.dna_rm.chr4.intergenic_gt122.orf:1...          3  \n",
       "GRCh38.Ens89.dna_rm.chr5.intergenic_gt122.orf:1...          3  \n",
       "GRCh38.Ens89.dna_rm.chr7.intergenic_gt122.orf:5...          3  \n",
       "GRCh38.Ens89.dna_rm.chr9.intergenic_gt122.orf:6...          3  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[~df2.Status.str.contains('norf')].iloc[::2000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "db8e7cee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count check: 3943 + 3915 == 7858 ? True \n",
      "count check: 3947 + 3902 == 7849 ? True \n",
      "count check: 3929 + 3902 == 7831 ? True \n",
      "count check: 3924 + 3901 == 7825 ? True \n",
      "count check: 3911 + 3903 == 7814 ? True \n"
     ]
    }
   ],
   "source": [
    "# For each control set, write an output BED file for the ORF and Non-ORF sets\n",
    "\n",
    "for control_set in sorted(df2.Control_Set.unique()):\n",
    "    # print(control_set)\n",
    "    df_temp = df2[df2.Control_Set == control_set].sort_values(by=['Chr','Gene_Start_hg38','Gene_End_hg38'])\n",
    "    df_temp_orf = df_temp[~df_temp.Status.str.contains('norf')]\n",
    "    df_temp_norf = df_temp[df_temp.Status.str.contains('norf')]\n",
    "    print(f'count check: {df_temp_orf.shape[0]} + {df_temp_norf.shape[0]} == {df_temp.shape[0]} ? {df_temp_orf.shape[0] + df_temp_norf.shape[0] == df_temp.shape[0]} ')\n",
    "    df_by_set[control_set] = {'orf':bedify(df_temp_orf),'norf':bedify(df_temp_norf)}\n",
    "    bedify(df_temp_orf).to_csv(f'grch38_orf_{control_set}.bed',sep='\\t',index=False,header=False,encoding=\"utf-8\")\n",
    "    bedify(df_temp_norf).to_csv(f'grch38_norf_{control_set}.bed',sep='\\t',index=False,header=False,encoding=\"utf-8\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4535b439",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove heirarchy and write out removed sequences due to Annotated and Unannotated Genes\n",
    "pd.concat(ugag_removal_list).drop('heirarchy',axis=1).sort_values(by=[\"Chr\",'Gene_Start_hg38','Gene_End_hg38']).to_csv('removed_sequences_genic_overlap.tsv',sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4d761b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out sequences removed due to ORF-NORF overlap\n",
    "pd.concat(removed_list).drop('heirarchy',axis=1).sort_values(by=[\"Chr\",'Gene_Start_hg38','Gene_End_hg38']).to_csv('removed_sequences_orf_norf_overlap.tsv',sep='\\t')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "71e9653e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out ORFs removed due to duplication\n",
    "pd.concat(dupl_list).drop('heirarchy',axis=1).sort_values(by=[\"Chr\",'Gene_Start_hg38','Gene_End_hg38']).to_csv('removed_sequences_duplicates.tsv',sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3b7cb2c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25ebf709",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
